{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e5976b8e",
   "metadata": {},
   "source": [
    "# Inferring enhancer-driven Gene Regulatory Networks (eGRNs) using SCENIC+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e045568f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up Environment\n",
    "import dill\n",
    "import scanpy as sc\n",
    "import os\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import pandas\n",
    "import pyranges\n",
    "# Set stderr to null to avoid strange messages from ray\n",
    "import sys\n",
    "_stderr = sys.stderr\n",
    "null = open(os.devnull,'wb')\n",
    "\n",
    "# set working directory\n",
    "work_dir = '/g/scb/zaugg/deuner/SCENIC+/'\n",
    "# set tmp directory\n",
    "tmp_dir = '/g/scb/zaugg/deuner/SCENIC+/tmp/combined/'\n",
    "# set the figures directory\n",
    "fig_dir = '/g/scb/zaugg/deuner/SCENIC+/figures/'\n",
    "# set the output data directory\n",
    "out_dir = '/g/scb/zaugg/deuner/SCENIC+/outputdata/'\n",
    "\n",
    "# Load the AnnData object containing the scRNA-seq side of the analysis\n",
    "adata = sc.read_h5ad(os.path.join(tmp_dir, 'combined.nomicro.adata.h5ad'))\n",
    "\n",
    "# Load the cisTopic object containing the scATAC-seq side of the analysis.\n",
    "cistopic_obj = dill.load(open(os.path.join(tmp_dir, 'scATAC/cistopic_obj.pkl'), 'rb'))\n",
    "\n",
    "# Load the motif enrichment dictionary containing the motif enrichment results.\n",
    "menr = dill.load(open(os.path.join(tmp_dir, 'motifs/menr.pkl'), 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "162b6906-e032-4fc9-8b7c-38f81e0175f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dupl_rate</th>\n",
       "      <th>cisTopic_log_nr_acc</th>\n",
       "      <th>Total_nr_frag_in_regions</th>\n",
       "      <th>Log_total_nr_frag</th>\n",
       "      <th>Total_nr_frag</th>\n",
       "      <th>cisTopic_nr_frag</th>\n",
       "      <th>Unique_nr_frag_in_regions</th>\n",
       "      <th>Log_unique_nr_frag</th>\n",
       "      <th>cisTopic_nr_acc</th>\n",
       "      <th>cisTopic_log_nr_frag</th>\n",
       "      <th>...</th>\n",
       "      <th>wsnn_res.9</th>\n",
       "      <th>wsnn_res.10</th>\n",
       "      <th>wsnn_res.12</th>\n",
       "      <th>wsnn_res.14</th>\n",
       "      <th>wsnn_res.16</th>\n",
       "      <th>wsnn_res.18</th>\n",
       "      <th>wsnn_res.20</th>\n",
       "      <th>celltype_wnn</th>\n",
       "      <th>barcode</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCAATTGCTCCTGAGT_timecourse</th>\n",
       "      <td>0.992421</td>\n",
       "      <td>3.625621</td>\n",
       "      <td>470578</td>\n",
       "      <td>5.885086</td>\n",
       "      <td>767513</td>\n",
       "      <td>4379</td>\n",
       "      <td>3350</td>\n",
       "      <td>3.764699</td>\n",
       "      <td>4223</td>\n",
       "      <td>3.641375</td>\n",
       "      <td>...</td>\n",
       "      <td>47</td>\n",
       "      <td>49</td>\n",
       "      <td>50</td>\n",
       "      <td>47</td>\n",
       "      <td>34</td>\n",
       "      <td>147</td>\n",
       "      <td>151</td>\n",
       "      <td>hiPSC-2</td>\n",
       "      <td>TCAATTGCTCCTGAGT</td>\n",
       "      <td>timecourse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGAACCGACTATCCTG_timecourse</th>\n",
       "      <td>0.992839</td>\n",
       "      <td>3.804957</td>\n",
       "      <td>781384</td>\n",
       "      <td>6.076014</td>\n",
       "      <td>1191281</td>\n",
       "      <td>6794</td>\n",
       "      <td>5182</td>\n",
       "      <td>3.931000</td>\n",
       "      <td>6382</td>\n",
       "      <td>3.832126</td>\n",
       "      <td>...</td>\n",
       "      <td>34</td>\n",
       "      <td>32</td>\n",
       "      <td>26</td>\n",
       "      <td>23</td>\n",
       "      <td>14</td>\n",
       "      <td>62</td>\n",
       "      <td>50</td>\n",
       "      <td>diff-neuron</td>\n",
       "      <td>TGAACCGACTATCCTG</td>\n",
       "      <td>timecourse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGATTAGTGGCTTTAA_timecourse</th>\n",
       "      <td>0.992924</td>\n",
       "      <td>4.093352</td>\n",
       "      <td>1642543</td>\n",
       "      <td>6.460567</td>\n",
       "      <td>2887798</td>\n",
       "      <td>13338</td>\n",
       "      <td>11197</td>\n",
       "      <td>4.310375</td>\n",
       "      <td>12398</td>\n",
       "      <td>4.125091</td>\n",
       "      <td>...</td>\n",
       "      <td>55</td>\n",
       "      <td>40</td>\n",
       "      <td>8</td>\n",
       "      <td>61</td>\n",
       "      <td>50</td>\n",
       "      <td>36</td>\n",
       "      <td>21</td>\n",
       "      <td>hiPSC-1</td>\n",
       "      <td>GGATTAGTGGCTTTAA</td>\n",
       "      <td>timecourse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAGCCTAACTAATCAG_timecourse</th>\n",
       "      <td>0.992291</td>\n",
       "      <td>4.273441</td>\n",
       "      <td>2496161</td>\n",
       "      <td>6.590419</td>\n",
       "      <td>3894208</td>\n",
       "      <td>20864</td>\n",
       "      <td>19777</td>\n",
       "      <td>4.477411</td>\n",
       "      <td>18769</td>\n",
       "      <td>4.319398</td>\n",
       "      <td>...</td>\n",
       "      <td>50</td>\n",
       "      <td>97</td>\n",
       "      <td>105</td>\n",
       "      <td>113</td>\n",
       "      <td>114</td>\n",
       "      <td>125</td>\n",
       "      <td>115</td>\n",
       "      <td>hiPSC-2</td>\n",
       "      <td>AAGCCTAACTAATCAG</td>\n",
       "      <td>timecourse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGCGGTTACCTGGATG_timecourse</th>\n",
       "      <td>0.992954</td>\n",
       "      <td>3.723948</td>\n",
       "      <td>655078</td>\n",
       "      <td>6.012438</td>\n",
       "      <td>1029054</td>\n",
       "      <td>5564</td>\n",
       "      <td>4457</td>\n",
       "      <td>3.860398</td>\n",
       "      <td>5296</td>\n",
       "      <td>3.745387</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>37</td>\n",
       "      <td>22</td>\n",
       "      <td>70</td>\n",
       "      <td>48</td>\n",
       "      <td>143</td>\n",
       "      <td>156</td>\n",
       "      <td>diff-NPC</td>\n",
       "      <td>GGCGGTTACCTGGATG</td>\n",
       "      <td>timecourse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GATGAAGGAGGCCCTT_cocultured28</th>\n",
       "      <td>0.993454</td>\n",
       "      <td>3.123525</td>\n",
       "      <td>190524</td>\n",
       "      <td>5.496168</td>\n",
       "      <td>313450</td>\n",
       "      <td>1379</td>\n",
       "      <td>1164</td>\n",
       "      <td>3.312177</td>\n",
       "      <td>1329</td>\n",
       "      <td>3.139564</td>\n",
       "      <td>...</td>\n",
       "      <td>104</td>\n",
       "      <td>108</td>\n",
       "      <td>127</td>\n",
       "      <td>132</td>\n",
       "      <td>148</td>\n",
       "      <td>157</td>\n",
       "      <td>164</td>\n",
       "      <td>neuron-3</td>\n",
       "      <td>GATGAAGGAGGCCCTT</td>\n",
       "      <td>cocultured28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTGTATTGATCCTAAT_cocultured28</th>\n",
       "      <td>0.992931</td>\n",
       "      <td>3.394802</td>\n",
       "      <td>336729</td>\n",
       "      <td>5.666258</td>\n",
       "      <td>463722</td>\n",
       "      <td>2584</td>\n",
       "      <td>2166</td>\n",
       "      <td>3.515609</td>\n",
       "      <td>2482</td>\n",
       "      <td>3.412293</td>\n",
       "      <td>...</td>\n",
       "      <td>33</td>\n",
       "      <td>43</td>\n",
       "      <td>29</td>\n",
       "      <td>21</td>\n",
       "      <td>45</td>\n",
       "      <td>73</td>\n",
       "      <td>48</td>\n",
       "      <td>neuron-3</td>\n",
       "      <td>CTGTATTGATCCTAAT</td>\n",
       "      <td>cocultured28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCTAAGCCTTGAGGTA_cocultured28</th>\n",
       "      <td>0.993037</td>\n",
       "      <td>3.474216</td>\n",
       "      <td>413668</td>\n",
       "      <td>5.762289</td>\n",
       "      <td>578481</td>\n",
       "      <td>3171</td>\n",
       "      <td>2706</td>\n",
       "      <td>3.605089</td>\n",
       "      <td>2980</td>\n",
       "      <td>3.501196</td>\n",
       "      <td>...</td>\n",
       "      <td>22</td>\n",
       "      <td>14</td>\n",
       "      <td>55</td>\n",
       "      <td>49</td>\n",
       "      <td>94</td>\n",
       "      <td>80</td>\n",
       "      <td>90</td>\n",
       "      <td>neuron-2</td>\n",
       "      <td>CCTAAGCCTTGAGGTA</td>\n",
       "      <td>cocultured28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCCACAATGCGTGATT_cocultured28</th>\n",
       "      <td>0.993336</td>\n",
       "      <td>3.218273</td>\n",
       "      <td>220009</td>\n",
       "      <td>5.499695</td>\n",
       "      <td>316006</td>\n",
       "      <td>1693</td>\n",
       "      <td>1345</td>\n",
       "      <td>3.323458</td>\n",
       "      <td>1653</td>\n",
       "      <td>3.228657</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>61</td>\n",
       "      <td>62</td>\n",
       "      <td>57</td>\n",
       "      <td>134</td>\n",
       "      <td>129</td>\n",
       "      <td>136</td>\n",
       "      <td>neuron-3</td>\n",
       "      <td>TCCACAATGCGTGATT</td>\n",
       "      <td>cocultured28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCGAGTGCAAGCTT_cocultured28</th>\n",
       "      <td>0.992965</td>\n",
       "      <td>3.186956</td>\n",
       "      <td>196610</td>\n",
       "      <td>5.469131</td>\n",
       "      <td>294531</td>\n",
       "      <td>1590</td>\n",
       "      <td>1268</td>\n",
       "      <td>3.316390</td>\n",
       "      <td>1538</td>\n",
       "      <td>3.201397</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>61</td>\n",
       "      <td>62</td>\n",
       "      <td>57</td>\n",
       "      <td>142</td>\n",
       "      <td>152</td>\n",
       "      <td>123</td>\n",
       "      <td>neuron-3</td>\n",
       "      <td>GTGCGAGTGCAAGCTT</td>\n",
       "      <td>cocultured28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19827 rows × 69 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               Dupl_rate cisTopic_log_nr_acc  \\\n",
       "TCAATTGCTCCTGAGT_timecourse     0.992421            3.625621   \n",
       "TGAACCGACTATCCTG_timecourse     0.992839            3.804957   \n",
       "GGATTAGTGGCTTTAA_timecourse     0.992924            4.093352   \n",
       "AAGCCTAACTAATCAG_timecourse     0.992291            4.273441   \n",
       "GGCGGTTACCTGGATG_timecourse     0.992954            3.723948   \n",
       "...                                  ...                 ...   \n",
       "GATGAAGGAGGCCCTT_cocultured28   0.993454            3.123525   \n",
       "CTGTATTGATCCTAAT_cocultured28   0.992931            3.394802   \n",
       "CCTAAGCCTTGAGGTA_cocultured28   0.993037            3.474216   \n",
       "TCCACAATGCGTGATT_cocultured28   0.993336            3.218273   \n",
       "GTGCGAGTGCAAGCTT_cocultured28   0.992965            3.186956   \n",
       "\n",
       "                               Total_nr_frag_in_regions  Log_total_nr_frag  \\\n",
       "TCAATTGCTCCTGAGT_timecourse                      470578           5.885086   \n",
       "TGAACCGACTATCCTG_timecourse                      781384           6.076014   \n",
       "GGATTAGTGGCTTTAA_timecourse                     1642543           6.460567   \n",
       "AAGCCTAACTAATCAG_timecourse                     2496161           6.590419   \n",
       "GGCGGTTACCTGGATG_timecourse                      655078           6.012438   \n",
       "...                                                 ...                ...   \n",
       "GATGAAGGAGGCCCTT_cocultured28                    190524           5.496168   \n",
       "CTGTATTGATCCTAAT_cocultured28                    336729           5.666258   \n",
       "CCTAAGCCTTGAGGTA_cocultured28                    413668           5.762289   \n",
       "TCCACAATGCGTGATT_cocultured28                    220009           5.499695   \n",
       "GTGCGAGTGCAAGCTT_cocultured28                    196610           5.469131   \n",
       "\n",
       "                               Total_nr_frag cisTopic_nr_frag  \\\n",
       "TCAATTGCTCCTGAGT_timecourse           767513             4379   \n",
       "TGAACCGACTATCCTG_timecourse          1191281             6794   \n",
       "GGATTAGTGGCTTTAA_timecourse          2887798            13338   \n",
       "AAGCCTAACTAATCAG_timecourse          3894208            20864   \n",
       "GGCGGTTACCTGGATG_timecourse          1029054             5564   \n",
       "...                                      ...              ...   \n",
       "GATGAAGGAGGCCCTT_cocultured28         313450             1379   \n",
       "CTGTATTGATCCTAAT_cocultured28         463722             2584   \n",
       "CCTAAGCCTTGAGGTA_cocultured28         578481             3171   \n",
       "TCCACAATGCGTGATT_cocultured28         316006             1693   \n",
       "GTGCGAGTGCAAGCTT_cocultured28         294531             1590   \n",
       "\n",
       "                               Unique_nr_frag_in_regions  Log_unique_nr_frag  \\\n",
       "TCAATTGCTCCTGAGT_timecourse                         3350            3.764699   \n",
       "TGAACCGACTATCCTG_timecourse                         5182            3.931000   \n",
       "GGATTAGTGGCTTTAA_timecourse                        11197            4.310375   \n",
       "AAGCCTAACTAATCAG_timecourse                        19777            4.477411   \n",
       "GGCGGTTACCTGGATG_timecourse                         4457            3.860398   \n",
       "...                                                  ...                 ...   \n",
       "GATGAAGGAGGCCCTT_cocultured28                       1164            3.312177   \n",
       "CTGTATTGATCCTAAT_cocultured28                       2166            3.515609   \n",
       "CCTAAGCCTTGAGGTA_cocultured28                       2706            3.605089   \n",
       "TCCACAATGCGTGATT_cocultured28                       1345            3.323458   \n",
       "GTGCGAGTGCAAGCTT_cocultured28                       1268            3.316390   \n",
       "\n",
       "                              cisTopic_nr_acc cisTopic_log_nr_frag  ...  \\\n",
       "TCAATTGCTCCTGAGT_timecourse              4223             3.641375  ...   \n",
       "TGAACCGACTATCCTG_timecourse              6382             3.832126  ...   \n",
       "GGATTAGTGGCTTTAA_timecourse             12398             4.125091  ...   \n",
       "AAGCCTAACTAATCAG_timecourse             18769             4.319398  ...   \n",
       "GGCGGTTACCTGGATG_timecourse              5296             3.745387  ...   \n",
       "...                                       ...                  ...  ...   \n",
       "GATGAAGGAGGCCCTT_cocultured28            1329             3.139564  ...   \n",
       "CTGTATTGATCCTAAT_cocultured28            2482             3.412293  ...   \n",
       "CCTAAGCCTTGAGGTA_cocultured28            2980             3.501196  ...   \n",
       "TCCACAATGCGTGATT_cocultured28            1653             3.228657  ...   \n",
       "GTGCGAGTGCAAGCTT_cocultured28            1538             3.201397  ...   \n",
       "\n",
       "                               wsnn_res.9  wsnn_res.10  wsnn_res.12  \\\n",
       "TCAATTGCTCCTGAGT_timecourse            47           49           50   \n",
       "TGAACCGACTATCCTG_timecourse            34           32           26   \n",
       "GGATTAGTGGCTTTAA_timecourse            55           40            8   \n",
       "AAGCCTAACTAATCAG_timecourse            50           97          105   \n",
       "GGCGGTTACCTGGATG_timecourse             8           37           22   \n",
       "...                                   ...          ...          ...   \n",
       "GATGAAGGAGGCCCTT_cocultured28         104          108          127   \n",
       "CTGTATTGATCCTAAT_cocultured28          33           43           29   \n",
       "CCTAAGCCTTGAGGTA_cocultured28          22           14           55   \n",
       "TCCACAATGCGTGATT_cocultured28          15           61           62   \n",
       "GTGCGAGTGCAAGCTT_cocultured28          15           61           62   \n",
       "\n",
       "                               wsnn_res.14 wsnn_res.16  wsnn_res.18  \\\n",
       "TCAATTGCTCCTGAGT_timecourse             47          34          147   \n",
       "TGAACCGACTATCCTG_timecourse             23          14           62   \n",
       "GGATTAGTGGCTTTAA_timecourse             61          50           36   \n",
       "AAGCCTAACTAATCAG_timecourse            113         114          125   \n",
       "GGCGGTTACCTGGATG_timecourse             70          48          143   \n",
       "...                                    ...         ...          ...   \n",
       "GATGAAGGAGGCCCTT_cocultured28          132         148          157   \n",
       "CTGTATTGATCCTAAT_cocultured28           21          45           73   \n",
       "CCTAAGCCTTGAGGTA_cocultured28           49          94           80   \n",
       "TCCACAATGCGTGATT_cocultured28           57         134          129   \n",
       "GTGCGAGTGCAAGCTT_cocultured28           57         142          152   \n",
       "\n",
       "                               wsnn_res.20  celltype_wnn           barcode  \\\n",
       "TCAATTGCTCCTGAGT_timecourse            151       hiPSC-2  TCAATTGCTCCTGAGT   \n",
       "TGAACCGACTATCCTG_timecourse             50   diff-neuron  TGAACCGACTATCCTG   \n",
       "GGATTAGTGGCTTTAA_timecourse             21       hiPSC-1  GGATTAGTGGCTTTAA   \n",
       "AAGCCTAACTAATCAG_timecourse            115       hiPSC-2  AAGCCTAACTAATCAG   \n",
       "GGCGGTTACCTGGATG_timecourse            156      diff-NPC  GGCGGTTACCTGGATG   \n",
       "...                                    ...           ...               ...   \n",
       "GATGAAGGAGGCCCTT_cocultured28          164      neuron-3  GATGAAGGAGGCCCTT   \n",
       "CTGTATTGATCCTAAT_cocultured28           48      neuron-3  CTGTATTGATCCTAAT   \n",
       "CCTAAGCCTTGAGGTA_cocultured28           90      neuron-2  CCTAAGCCTTGAGGTA   \n",
       "TCCACAATGCGTGATT_cocultured28          136      neuron-3  TCCACAATGCGTGATT   \n",
       "GTGCGAGTGCAAGCTT_cocultured28          123      neuron-3  GTGCGAGTGCAAGCTT   \n",
       "\n",
       "                                  sample_id  \n",
       "TCAATTGCTCCTGAGT_timecourse      timecourse  \n",
       "TGAACCGACTATCCTG_timecourse      timecourse  \n",
       "GGATTAGTGGCTTTAA_timecourse      timecourse  \n",
       "AAGCCTAACTAATCAG_timecourse      timecourse  \n",
       "GGCGGTTACCTGGATG_timecourse      timecourse  \n",
       "...                                     ...  \n",
       "GATGAAGGAGGCCCTT_cocultured28  cocultured28  \n",
       "CTGTATTGATCCTAAT_cocultured28  cocultured28  \n",
       "CCTAAGCCTTGAGGTA_cocultured28  cocultured28  \n",
       "TCCACAATGCGTGATT_cocultured28  cocultured28  \n",
       "GTGCGAGTGCAAGCTT_cocultured28  cocultured28  \n",
       "\n",
       "[19827 rows x 69 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cistopic_obj.cell_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "88c2e311-7799-436e-a1e2-70c8cde09ff5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adapt barcodes of cistopic object\n",
    "new_bcs = []\n",
    "old_bcs = cistopic_obj.cell_names\n",
    "for i in range(len(old_bcs)):\n",
    "    split_bc = str.split(old_bcs[i], \"_\")\n",
    "    new_bc = split_bc[1] + \"_\" + split_bc[0]\n",
    "    new_bcs.append(new_bc)\n",
    "    cistopic_obj.selected_model.cell_topic.columns.values[i] = new_bc\n",
    "\n",
    "cistopic_obj.cell_names = new_bcs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5379e121-b625-4513-b8ab-7103cad1434d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cistopic_obj.cell_data.index = new_bcs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d82563d9-96b1-49d2-ac36-01eaff718e48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>timecourse_TCAATTGCTCCTGAGT</th>\n",
       "      <th>timecourse_TGAACCGACTATCCTG</th>\n",
       "      <th>timecourse_GGATTAGTGGCTTTAA</th>\n",
       "      <th>timecourse_AAGCCTAACTAATCAG</th>\n",
       "      <th>timecourse_GGCGGTTACCTGGATG</th>\n",
       "      <th>timecourse_CGTTAAGGAGTCAATA</th>\n",
       "      <th>timecourse_GATAGCCGACCTTTGG</th>\n",
       "      <th>timecourse_CACCACACTATCCTCC</th>\n",
       "      <th>timecourse_GATTTGGACCCACATT</th>\n",
       "      <th>timecourse_CCAACCTGACCGTTGT</th>\n",
       "      <th>...</th>\n",
       "      <th>cocultured28_AGCAATACTACATAAG</th>\n",
       "      <th>cocultured28_AGGAAGCACAATTGGC</th>\n",
       "      <th>cocultured28_TGAATTGACTTCATCG</th>\n",
       "      <th>cocultured28_CCACTATACCTAGTCT</th>\n",
       "      <th>cocultured28_ATGGCTAACGCTTGCT</th>\n",
       "      <th>cocultured28_GATGAAGGAGGCCCTT</th>\n",
       "      <th>cocultured28_CTGTATTGATCCTAAT</th>\n",
       "      <th>cocultured28_CCTAAGCCTTGAGGTA</th>\n",
       "      <th>cocultured28_TCCACAATGCGTGATT</th>\n",
       "      <th>cocultured28_GTGCGAGTGCAAGCTT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Topic1</th>\n",
       "      <td>0.012872</td>\n",
       "      <td>0.448228</td>\n",
       "      <td>0.022895</td>\n",
       "      <td>0.016632</td>\n",
       "      <td>0.392630</td>\n",
       "      <td>0.026084</td>\n",
       "      <td>0.136395</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.427873</td>\n",
       "      <td>0.222501</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009130</td>\n",
       "      <td>0.034632</td>\n",
       "      <td>0.055914</td>\n",
       "      <td>0.020237</td>\n",
       "      <td>0.050934</td>\n",
       "      <td>0.094996</td>\n",
       "      <td>0.041469</td>\n",
       "      <td>0.011221</td>\n",
       "      <td>0.039930</td>\n",
       "      <td>0.019521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic2</th>\n",
       "      <td>0.014978</td>\n",
       "      <td>0.040112</td>\n",
       "      <td>0.035749</td>\n",
       "      <td>0.016526</td>\n",
       "      <td>0.285073</td>\n",
       "      <td>0.021246</td>\n",
       "      <td>0.023561</td>\n",
       "      <td>0.248934</td>\n",
       "      <td>0.359294</td>\n",
       "      <td>0.453219</td>\n",
       "      <td>...</td>\n",
       "      <td>0.021144</td>\n",
       "      <td>0.010462</td>\n",
       "      <td>0.012289</td>\n",
       "      <td>0.011455</td>\n",
       "      <td>0.026599</td>\n",
       "      <td>0.019579</td>\n",
       "      <td>0.019747</td>\n",
       "      <td>0.012871</td>\n",
       "      <td>0.039930</td>\n",
       "      <td>0.034635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic3</th>\n",
       "      <td>0.008893</td>\n",
       "      <td>0.008240</td>\n",
       "      <td>0.024100</td>\n",
       "      <td>0.029332</td>\n",
       "      <td>0.025065</td>\n",
       "      <td>0.015081</td>\n",
       "      <td>0.011223</td>\n",
       "      <td>0.025586</td>\n",
       "      <td>0.014977</td>\n",
       "      <td>0.009376</td>\n",
       "      <td>...</td>\n",
       "      <td>0.013455</td>\n",
       "      <td>0.008658</td>\n",
       "      <td>0.019969</td>\n",
       "      <td>0.012600</td>\n",
       "      <td>0.016412</td>\n",
       "      <td>0.035533</td>\n",
       "      <td>0.009084</td>\n",
       "      <td>0.009241</td>\n",
       "      <td>0.012331</td>\n",
       "      <td>0.015743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic4</th>\n",
       "      <td>0.628130</td>\n",
       "      <td>0.006063</td>\n",
       "      <td>0.539444</td>\n",
       "      <td>0.622403</td>\n",
       "      <td>0.034979</td>\n",
       "      <td>0.015650</td>\n",
       "      <td>0.011138</td>\n",
       "      <td>0.031983</td>\n",
       "      <td>0.008671</td>\n",
       "      <td>0.056157</td>\n",
       "      <td>...</td>\n",
       "      <td>0.024027</td>\n",
       "      <td>0.009019</td>\n",
       "      <td>0.017512</td>\n",
       "      <td>0.019855</td>\n",
       "      <td>0.019808</td>\n",
       "      <td>0.042059</td>\n",
       "      <td>0.026856</td>\n",
       "      <td>0.017822</td>\n",
       "      <td>0.018203</td>\n",
       "      <td>0.018262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic5</th>\n",
       "      <td>0.078399</td>\n",
       "      <td>0.091573</td>\n",
       "      <td>0.194489</td>\n",
       "      <td>0.182209</td>\n",
       "      <td>0.139731</td>\n",
       "      <td>0.139144</td>\n",
       "      <td>0.110093</td>\n",
       "      <td>0.086887</td>\n",
       "      <td>0.091597</td>\n",
       "      <td>0.164315</td>\n",
       "      <td>...</td>\n",
       "      <td>0.060067</td>\n",
       "      <td>0.040043</td>\n",
       "      <td>0.022427</td>\n",
       "      <td>0.112257</td>\n",
       "      <td>0.020374</td>\n",
       "      <td>0.108049</td>\n",
       "      <td>0.032385</td>\n",
       "      <td>0.024092</td>\n",
       "      <td>0.056371</td>\n",
       "      <td>0.032746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic6</th>\n",
       "      <td>0.065996</td>\n",
       "      <td>0.024720</td>\n",
       "      <td>0.112870</td>\n",
       "      <td>0.075774</td>\n",
       "      <td>0.041152</td>\n",
       "      <td>0.062790</td>\n",
       "      <td>0.036155</td>\n",
       "      <td>0.049574</td>\n",
       "      <td>0.041621</td>\n",
       "      <td>0.021748</td>\n",
       "      <td>...</td>\n",
       "      <td>0.037963</td>\n",
       "      <td>0.008658</td>\n",
       "      <td>0.003687</td>\n",
       "      <td>0.019855</td>\n",
       "      <td>0.021505</td>\n",
       "      <td>0.060914</td>\n",
       "      <td>0.025276</td>\n",
       "      <td>0.013531</td>\n",
       "      <td>0.032296</td>\n",
       "      <td>0.005668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic7</th>\n",
       "      <td>0.051252</td>\n",
       "      <td>0.004509</td>\n",
       "      <td>0.040167</td>\n",
       "      <td>0.036612</td>\n",
       "      <td>0.008792</td>\n",
       "      <td>0.008821</td>\n",
       "      <td>0.009853</td>\n",
       "      <td>0.015458</td>\n",
       "      <td>0.012770</td>\n",
       "      <td>0.018848</td>\n",
       "      <td>...</td>\n",
       "      <td>0.018260</td>\n",
       "      <td>0.027417</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.009164</td>\n",
       "      <td>0.022071</td>\n",
       "      <td>0.020305</td>\n",
       "      <td>0.015403</td>\n",
       "      <td>0.010891</td>\n",
       "      <td>0.028773</td>\n",
       "      <td>0.030227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic8</th>\n",
       "      <td>0.121226</td>\n",
       "      <td>0.168532</td>\n",
       "      <td>0.009399</td>\n",
       "      <td>0.006908</td>\n",
       "      <td>0.066218</td>\n",
       "      <td>0.349331</td>\n",
       "      <td>0.258482</td>\n",
       "      <td>0.098081</td>\n",
       "      <td>0.030427</td>\n",
       "      <td>0.043882</td>\n",
       "      <td>...</td>\n",
       "      <td>0.676117</td>\n",
       "      <td>0.691919</td>\n",
       "      <td>0.686022</td>\n",
       "      <td>0.639939</td>\n",
       "      <td>0.441426</td>\n",
       "      <td>0.459028</td>\n",
       "      <td>0.425750</td>\n",
       "      <td>0.728053</td>\n",
       "      <td>0.410452</td>\n",
       "      <td>0.518892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic9</th>\n",
       "      <td>0.010765</td>\n",
       "      <td>0.165578</td>\n",
       "      <td>0.010925</td>\n",
       "      <td>0.009299</td>\n",
       "      <td>0.003928</td>\n",
       "      <td>0.012330</td>\n",
       "      <td>0.037611</td>\n",
       "      <td>0.007996</td>\n",
       "      <td>0.004572</td>\n",
       "      <td>0.004833</td>\n",
       "      <td>...</td>\n",
       "      <td>0.095147</td>\n",
       "      <td>0.123737</td>\n",
       "      <td>0.132719</td>\n",
       "      <td>0.086292</td>\n",
       "      <td>0.015846</td>\n",
       "      <td>0.038434</td>\n",
       "      <td>0.006319</td>\n",
       "      <td>0.077228</td>\n",
       "      <td>0.019378</td>\n",
       "      <td>0.013224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic10</th>\n",
       "      <td>0.007489</td>\n",
       "      <td>0.042444</td>\n",
       "      <td>0.009961</td>\n",
       "      <td>0.004304</td>\n",
       "      <td>0.002432</td>\n",
       "      <td>0.349521</td>\n",
       "      <td>0.365490</td>\n",
       "      <td>0.006930</td>\n",
       "      <td>0.008198</td>\n",
       "      <td>0.005123</td>\n",
       "      <td>...</td>\n",
       "      <td>0.044690</td>\n",
       "      <td>0.045455</td>\n",
       "      <td>0.020891</td>\n",
       "      <td>0.068347</td>\n",
       "      <td>0.365025</td>\n",
       "      <td>0.121102</td>\n",
       "      <td>0.397709</td>\n",
       "      <td>0.095050</td>\n",
       "      <td>0.342337</td>\n",
       "      <td>0.311083</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 19827 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         timecourse_TCAATTGCTCCTGAGT  timecourse_TGAACCGACTATCCTG  \\\n",
       "Topic1                      0.012872                     0.448228   \n",
       "Topic2                      0.014978                     0.040112   \n",
       "Topic3                      0.008893                     0.008240   \n",
       "Topic4                      0.628130                     0.006063   \n",
       "Topic5                      0.078399                     0.091573   \n",
       "Topic6                      0.065996                     0.024720   \n",
       "Topic7                      0.051252                     0.004509   \n",
       "Topic8                      0.121226                     0.168532   \n",
       "Topic9                      0.010765                     0.165578   \n",
       "Topic10                     0.007489                     0.042444   \n",
       "\n",
       "         timecourse_GGATTAGTGGCTTTAA  timecourse_AAGCCTAACTAATCAG  \\\n",
       "Topic1                      0.022895                     0.016632   \n",
       "Topic2                      0.035749                     0.016526   \n",
       "Topic3                      0.024100                     0.029332   \n",
       "Topic4                      0.539444                     0.622403   \n",
       "Topic5                      0.194489                     0.182209   \n",
       "Topic6                      0.112870                     0.075774   \n",
       "Topic7                      0.040167                     0.036612   \n",
       "Topic8                      0.009399                     0.006908   \n",
       "Topic9                      0.010925                     0.009299   \n",
       "Topic10                     0.009961                     0.004304   \n",
       "\n",
       "         timecourse_GGCGGTTACCTGGATG  timecourse_CGTTAAGGAGTCAATA  \\\n",
       "Topic1                      0.392630                     0.026084   \n",
       "Topic2                      0.285073                     0.021246   \n",
       "Topic3                      0.025065                     0.015081   \n",
       "Topic4                      0.034979                     0.015650   \n",
       "Topic5                      0.139731                     0.139144   \n",
       "Topic6                      0.041152                     0.062790   \n",
       "Topic7                      0.008792                     0.008821   \n",
       "Topic8                      0.066218                     0.349331   \n",
       "Topic9                      0.003928                     0.012330   \n",
       "Topic10                     0.002432                     0.349521   \n",
       "\n",
       "         timecourse_GATAGCCGACCTTTGG  timecourse_CACCACACTATCCTCC  \\\n",
       "Topic1                      0.136395                     0.428571   \n",
       "Topic2                      0.023561                     0.248934   \n",
       "Topic3                      0.011223                     0.025586   \n",
       "Topic4                      0.011138                     0.031983   \n",
       "Topic5                      0.110093                     0.086887   \n",
       "Topic6                      0.036155                     0.049574   \n",
       "Topic7                      0.009853                     0.015458   \n",
       "Topic8                      0.258482                     0.098081   \n",
       "Topic9                      0.037611                     0.007996   \n",
       "Topic10                     0.365490                     0.006930   \n",
       "\n",
       "         timecourse_GATTTGGACCCACATT  timecourse_CCAACCTGACCGTTGT  ...  \\\n",
       "Topic1                      0.427873                     0.222501  ...   \n",
       "Topic2                      0.359294                     0.453219  ...   \n",
       "Topic3                      0.014977                     0.009376  ...   \n",
       "Topic4                      0.008671                     0.056157  ...   \n",
       "Topic5                      0.091597                     0.164315  ...   \n",
       "Topic6                      0.041621                     0.021748  ...   \n",
       "Topic7                      0.012770                     0.018848  ...   \n",
       "Topic8                      0.030427                     0.043882  ...   \n",
       "Topic9                      0.004572                     0.004833  ...   \n",
       "Topic10                     0.008198                     0.005123  ...   \n",
       "\n",
       "         cocultured28_AGCAATACTACATAAG  cocultured28_AGGAAGCACAATTGGC  \\\n",
       "Topic1                        0.009130                       0.034632   \n",
       "Topic2                        0.021144                       0.010462   \n",
       "Topic3                        0.013455                       0.008658   \n",
       "Topic4                        0.024027                       0.009019   \n",
       "Topic5                        0.060067                       0.040043   \n",
       "Topic6                        0.037963                       0.008658   \n",
       "Topic7                        0.018260                       0.027417   \n",
       "Topic8                        0.676117                       0.691919   \n",
       "Topic9                        0.095147                       0.123737   \n",
       "Topic10                       0.044690                       0.045455   \n",
       "\n",
       "         cocultured28_TGAATTGACTTCATCG  cocultured28_CCACTATACCTAGTCT  \\\n",
       "Topic1                        0.055914                       0.020237   \n",
       "Topic2                        0.012289                       0.011455   \n",
       "Topic3                        0.019969                       0.012600   \n",
       "Topic4                        0.017512                       0.019855   \n",
       "Topic5                        0.022427                       0.112257   \n",
       "Topic6                        0.003687                       0.019855   \n",
       "Topic7                        0.028571                       0.009164   \n",
       "Topic8                        0.686022                       0.639939   \n",
       "Topic9                        0.132719                       0.086292   \n",
       "Topic10                       0.020891                       0.068347   \n",
       "\n",
       "         cocultured28_ATGGCTAACGCTTGCT  cocultured28_GATGAAGGAGGCCCTT  \\\n",
       "Topic1                        0.050934                       0.094996   \n",
       "Topic2                        0.026599                       0.019579   \n",
       "Topic3                        0.016412                       0.035533   \n",
       "Topic4                        0.019808                       0.042059   \n",
       "Topic5                        0.020374                       0.108049   \n",
       "Topic6                        0.021505                       0.060914   \n",
       "Topic7                        0.022071                       0.020305   \n",
       "Topic8                        0.441426                       0.459028   \n",
       "Topic9                        0.015846                       0.038434   \n",
       "Topic10                       0.365025                       0.121102   \n",
       "\n",
       "         cocultured28_CTGTATTGATCCTAAT  cocultured28_CCTAAGCCTTGAGGTA  \\\n",
       "Topic1                        0.041469                       0.011221   \n",
       "Topic2                        0.019747                       0.012871   \n",
       "Topic3                        0.009084                       0.009241   \n",
       "Topic4                        0.026856                       0.017822   \n",
       "Topic5                        0.032385                       0.024092   \n",
       "Topic6                        0.025276                       0.013531   \n",
       "Topic7                        0.015403                       0.010891   \n",
       "Topic8                        0.425750                       0.728053   \n",
       "Topic9                        0.006319                       0.077228   \n",
       "Topic10                       0.397709                       0.095050   \n",
       "\n",
       "         cocultured28_TCCACAATGCGTGATT  cocultured28_GTGCGAGTGCAAGCTT  \n",
       "Topic1                        0.039930                       0.019521  \n",
       "Topic2                        0.039930                       0.034635  \n",
       "Topic3                        0.012331                       0.015743  \n",
       "Topic4                        0.018203                       0.018262  \n",
       "Topic5                        0.056371                       0.032746  \n",
       "Topic6                        0.032296                       0.005668  \n",
       "Topic7                        0.028773                       0.030227  \n",
       "Topic8                        0.410452                       0.518892  \n",
       "Topic9                        0.019378                       0.013224  \n",
       "Topic10                       0.342337                       0.311083  \n",
       "\n",
       "[10 rows x 19827 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cistopic_obj.selected_model.cell_topic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f46937bd-033f-4084-a9e6-0d26baa9070d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['timecourse_AAACAGCCAGCCAGTT', 'timecourse_AAACAGCCAGGCGAGT',\n",
       "       'timecourse_AAACAGCCAGTAAAGC', 'timecourse_AAACAGCCATAAGTCT',\n",
       "       'timecourse_AAACAGCCATAGGCGA', 'timecourse_AAACAGCCATCACAGC',\n",
       "       'timecourse_AAACATGCAACTAGCC', 'timecourse_AAACATGCAAGACTCC',\n",
       "       'timecourse_AAACATGCAGCACCAT', 'timecourse_AAACATGCAGCATTAT',\n",
       "       ...\n",
       "       'cocultured28_TTTGTCTAGGCGCTTA', 'cocultured28_TTTGTCTAGTGCACGC',\n",
       "       'cocultured28_TTTGTGAAGCTTAGTA', 'cocultured28_TTTGTGAAGGGTGAAC',\n",
       "       'cocultured28_TTTGTGAAGGGTGGAT', 'cocultured28_TTTGTGAAGTCTATGA',\n",
       "       'cocultured28_TTTGTGGCAGCAAGTG', 'cocultured28_TTTGTGTTCTCGCCTG',\n",
       "       'cocultured28_TTTGTGTTCTTAGCCC', 'cocultured28_TTTGTGTTCTTGCAGG'],\n",
       "      dtype='object', length=22367)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs\n",
    "adata.obs_names.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6b8a6744-87bd-443f-bc0a-a5815aab3a84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(adata.obs_names)) == len(adata.obs_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "98e66b81-0cd1-4c0c-8745-a35b3b6f55fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check if there are common barcodes\n",
    "list(set(adata.obs_names.copy(deep=True)) & set(list(cistopic_obj.cell_names.copy())))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "651a3126-5294-42e1-bdb2-5b8a2d3c55b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(list(cistopic_obj.cell_names.copy()))) > 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f22a4c53-5478-46fb-85be-8fc310671067",
   "metadata": {},
   "outputs": [],
   "source": [
    "# maybe select the atac barcodes as defaults for adata\n",
    "list(set(adata.obs[\"barcode\"]) & set(list(cistopic_obj.cell_names.copy())))\n",
    "\n",
    "l_bcs = []\n",
    "\n",
    "for i in range(len(adata.obs_names)):\n",
    "     l_bc = adata.obs[\"orig.ident\"][i] + \"_\" + adata.obs[\"barcode\"][i]\n",
    "     l_bcs.append(l_bc)\n",
    "     \n",
    "adata.obs[\"long_barcode\"] = l_bcs\n",
    "adata.obs_names = list(adata.obs[\"long_barcode\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "20c00414-17ce-43c4-9aa6-3f7224399655",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>percent.ribo</th>\n",
       "      <th>nCount_SCT</th>\n",
       "      <th>nFeature_SCT</th>\n",
       "      <th>SCT_snn_res.0.5</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>pANN_0.25_0.005_794</th>\n",
       "      <th>...</th>\n",
       "      <th>wsnn_res.9</th>\n",
       "      <th>wsnn_res.10</th>\n",
       "      <th>wsnn_res.12</th>\n",
       "      <th>wsnn_res.14</th>\n",
       "      <th>wsnn_res.16</th>\n",
       "      <th>wsnn_res.18</th>\n",
       "      <th>wsnn_res.20</th>\n",
       "      <th>celltype_wnn</th>\n",
       "      <th>barcode</th>\n",
       "      <th>long_barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>timecourse_AAACCGTACCCGCTGT</th>\n",
       "      <td>timecourse</td>\n",
       "      <td>5020.0</td>\n",
       "      <td>2676</td>\n",
       "      <td>6.155378</td>\n",
       "      <td>3.705179</td>\n",
       "      <td>6037.0</td>\n",
       "      <td>2676</td>\n",
       "      <td>7</td>\n",
       "      <td>25</td>\n",
       "      <td>0.084507</td>\n",
       "      <td>...</td>\n",
       "      <td>48</td>\n",
       "      <td>44</td>\n",
       "      <td>31</td>\n",
       "      <td>27</td>\n",
       "      <td>51</td>\n",
       "      <td>35</td>\n",
       "      <td>25</td>\n",
       "      <td>diff-NPC</td>\n",
       "      <td>AAACCGTACCCGCTGT</td>\n",
       "      <td>timecourse_AAACCGTACCCGCTGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>timecourse_GTAGGTTACCCGCTGT</th>\n",
       "      <td>timecourse</td>\n",
       "      <td>4719.0</td>\n",
       "      <td>2179</td>\n",
       "      <td>0.190718</td>\n",
       "      <td>0.466200</td>\n",
       "      <td>6032.0</td>\n",
       "      <td>2178</td>\n",
       "      <td>14</td>\n",
       "      <td>117</td>\n",
       "      <td>0.239437</td>\n",
       "      <td>...</td>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "      <td>109</td>\n",
       "      <td>114</td>\n",
       "      <td>120</td>\n",
       "      <td>115</td>\n",
       "      <td>117</td>\n",
       "      <td>neuron-1</td>\n",
       "      <td>GTAGGTTACCCGCTGT</td>\n",
       "      <td>timecourse_GTAGGTTACCCGCTGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>timecourse_CTAGTAAACCCGCTGT</th>\n",
       "      <td>timecourse</td>\n",
       "      <td>3832.0</td>\n",
       "      <td>2004</td>\n",
       "      <td>0.287056</td>\n",
       "      <td>0.730689</td>\n",
       "      <td>5979.0</td>\n",
       "      <td>2008</td>\n",
       "      <td>12</td>\n",
       "      <td>50</td>\n",
       "      <td>0.183099</td>\n",
       "      <td>...</td>\n",
       "      <td>34</td>\n",
       "      <td>32</td>\n",
       "      <td>26</td>\n",
       "      <td>23</td>\n",
       "      <td>14</td>\n",
       "      <td>62</td>\n",
       "      <td>50</td>\n",
       "      <td>diff-neuron</td>\n",
       "      <td>CTAGTAAACCCGCTGT</td>\n",
       "      <td>timecourse_CTAGTAAACCCGCTGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>timecourse_TGGCATGACCCGCTGT</th>\n",
       "      <td>timecourse</td>\n",
       "      <td>7329.0</td>\n",
       "      <td>3535</td>\n",
       "      <td>1.159776</td>\n",
       "      <td>1.132487</td>\n",
       "      <td>7247.0</td>\n",
       "      <td>3535</td>\n",
       "      <td>8</td>\n",
       "      <td>151</td>\n",
       "      <td>0.154930</td>\n",
       "      <td>...</td>\n",
       "      <td>47</td>\n",
       "      <td>49</td>\n",
       "      <td>50</td>\n",
       "      <td>47</td>\n",
       "      <td>34</td>\n",
       "      <td>147</td>\n",
       "      <td>151</td>\n",
       "      <td>hiPSC-2</td>\n",
       "      <td>TGGCATGACCCGCTGT</td>\n",
       "      <td>timecourse_TGGCATGACCCGCTGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>timecourse_TTTGTGCACCCGCTGT</th>\n",
       "      <td>timecourse</td>\n",
       "      <td>4511.0</td>\n",
       "      <td>2253</td>\n",
       "      <td>7.337619</td>\n",
       "      <td>1.906451</td>\n",
       "      <td>6061.0</td>\n",
       "      <td>2253</td>\n",
       "      <td>7</td>\n",
       "      <td>51</td>\n",
       "      <td>0.140845</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>41</td>\n",
       "      <td>36</td>\n",
       "      <td>70</td>\n",
       "      <td>52</td>\n",
       "      <td>51</td>\n",
       "      <td>diff-NPC</td>\n",
       "      <td>TTTGTGCACCCGCTGT</td>\n",
       "      <td>timecourse_TTTGTGCACCCGCTGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cocultured28_AGCAATTGACACCTAC</th>\n",
       "      <td>cocultured28</td>\n",
       "      <td>11717.0</td>\n",
       "      <td>3893</td>\n",
       "      <td>0.162158</td>\n",
       "      <td>0.187761</td>\n",
       "      <td>7918.0</td>\n",
       "      <td>3746</td>\n",
       "      <td>5</td>\n",
       "      <td>72</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>22</td>\n",
       "      <td>14</td>\n",
       "      <td>55</td>\n",
       "      <td>49</td>\n",
       "      <td>107</td>\n",
       "      <td>105</td>\n",
       "      <td>72</td>\n",
       "      <td>neuron-2</td>\n",
       "      <td>AGCAATTGACACCTAC</td>\n",
       "      <td>cocultured28_AGCAATTGACACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cocultured28_TGATTAGACACCAAAC</th>\n",
       "      <td>cocultured28</td>\n",
       "      <td>6845.0</td>\n",
       "      <td>2998</td>\n",
       "      <td>0.613587</td>\n",
       "      <td>0.555150</td>\n",
       "      <td>6845.0</td>\n",
       "      <td>2998</td>\n",
       "      <td>14</td>\n",
       "      <td>85</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>88</td>\n",
       "      <td>89</td>\n",
       "      <td>100</td>\n",
       "      <td>103</td>\n",
       "      <td>105</td>\n",
       "      <td>93</td>\n",
       "      <td>85</td>\n",
       "      <td>neuron-4</td>\n",
       "      <td>TGATTAGACACCAAAC</td>\n",
       "      <td>cocultured28_TGATTAGACACCAAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cocultured28_GTTAAAGCTTGAGAAC</th>\n",
       "      <td>cocultured28</td>\n",
       "      <td>9557.0</td>\n",
       "      <td>3435</td>\n",
       "      <td>0.292979</td>\n",
       "      <td>0.282515</td>\n",
       "      <td>7697.0</td>\n",
       "      <td>3433</td>\n",
       "      <td>6</td>\n",
       "      <td>39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>70</td>\n",
       "      <td>79</td>\n",
       "      <td>78</td>\n",
       "      <td>73</td>\n",
       "      <td>50</td>\n",
       "      <td>39</td>\n",
       "      <td>neuron-3</td>\n",
       "      <td>GTTAAAGCTTGAGAAC</td>\n",
       "      <td>cocultured28_GTTAAAGCTTGAGAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cocultured28_ACTAAGACTTGAGAAC</th>\n",
       "      <td>cocultured28</td>\n",
       "      <td>1732.0</td>\n",
       "      <td>1112</td>\n",
       "      <td>2.078522</td>\n",
       "      <td>0.577367</td>\n",
       "      <td>5364.0</td>\n",
       "      <td>1614</td>\n",
       "      <td>6</td>\n",
       "      <td>123</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>33</td>\n",
       "      <td>61</td>\n",
       "      <td>29</td>\n",
       "      <td>21</td>\n",
       "      <td>45</td>\n",
       "      <td>73</td>\n",
       "      <td>123</td>\n",
       "      <td>neuron-3</td>\n",
       "      <td>ACTAAGACTTGAGAAC</td>\n",
       "      <td>cocultured28_ACTAAGACTTGAGAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cocultured28_AAGGAAGCTTGAGAAC</th>\n",
       "      <td>cocultured28</td>\n",
       "      <td>8505.0</td>\n",
       "      <td>3170</td>\n",
       "      <td>0.282187</td>\n",
       "      <td>0.435038</td>\n",
       "      <td>7565.0</td>\n",
       "      <td>3170</td>\n",
       "      <td>6</td>\n",
       "      <td>48</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>33</td>\n",
       "      <td>43</td>\n",
       "      <td>29</td>\n",
       "      <td>21</td>\n",
       "      <td>45</td>\n",
       "      <td>134</td>\n",
       "      <td>48</td>\n",
       "      <td>neuron-3</td>\n",
       "      <td>AAGGAAGCTTGAGAAC</td>\n",
       "      <td>cocultured28_AAGGAAGCTTGAGAAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22367 rows × 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 orig.ident  nCount_RNA  nFeature_RNA  \\\n",
       "timecourse_AAACCGTACCCGCTGT      timecourse      5020.0          2676   \n",
       "timecourse_GTAGGTTACCCGCTGT      timecourse      4719.0          2179   \n",
       "timecourse_CTAGTAAACCCGCTGT      timecourse      3832.0          2004   \n",
       "timecourse_TGGCATGACCCGCTGT      timecourse      7329.0          3535   \n",
       "timecourse_TTTGTGCACCCGCTGT      timecourse      4511.0          2253   \n",
       "...                                     ...         ...           ...   \n",
       "cocultured28_AGCAATTGACACCTAC  cocultured28     11717.0          3893   \n",
       "cocultured28_TGATTAGACACCAAAC  cocultured28      6845.0          2998   \n",
       "cocultured28_GTTAAAGCTTGAGAAC  cocultured28      9557.0          3435   \n",
       "cocultured28_ACTAAGACTTGAGAAC  cocultured28      1732.0          1112   \n",
       "cocultured28_AAGGAAGCTTGAGAAC  cocultured28      8505.0          3170   \n",
       "\n",
       "                               percent.mt  percent.ribo  nCount_SCT  \\\n",
       "timecourse_AAACCGTACCCGCTGT      6.155378      3.705179      6037.0   \n",
       "timecourse_GTAGGTTACCCGCTGT      0.190718      0.466200      6032.0   \n",
       "timecourse_CTAGTAAACCCGCTGT      0.287056      0.730689      5979.0   \n",
       "timecourse_TGGCATGACCCGCTGT      1.159776      1.132487      7247.0   \n",
       "timecourse_TTTGTGCACCCGCTGT      7.337619      1.906451      6061.0   \n",
       "...                                   ...           ...         ...   \n",
       "cocultured28_AGCAATTGACACCTAC    0.162158      0.187761      7918.0   \n",
       "cocultured28_TGATTAGACACCAAAC    0.613587      0.555150      6845.0   \n",
       "cocultured28_GTTAAAGCTTGAGAAC    0.292979      0.282515      7697.0   \n",
       "cocultured28_ACTAAGACTTGAGAAC    2.078522      0.577367      5364.0   \n",
       "cocultured28_AAGGAAGCTTGAGAAC    0.282187      0.435038      7565.0   \n",
       "\n",
       "                               nFeature_SCT SCT_snn_res.0.5 seurat_clusters  \\\n",
       "timecourse_AAACCGTACCCGCTGT            2676               7              25   \n",
       "timecourse_GTAGGTTACCCGCTGT            2178              14             117   \n",
       "timecourse_CTAGTAAACCCGCTGT            2008              12              50   \n",
       "timecourse_TGGCATGACCCGCTGT            3535               8             151   \n",
       "timecourse_TTTGTGCACCCGCTGT            2253               7              51   \n",
       "...                                     ...             ...             ...   \n",
       "cocultured28_AGCAATTGACACCTAC          3746               5              72   \n",
       "cocultured28_TGATTAGACACCAAAC          2998              14              85   \n",
       "cocultured28_GTTAAAGCTTGAGAAC          3433               6              39   \n",
       "cocultured28_ACTAAGACTTGAGAAC          1614               6             123   \n",
       "cocultured28_AAGGAAGCTTGAGAAC          3170               6              48   \n",
       "\n",
       "                               pANN_0.25_0.005_794  ... wsnn_res.9  \\\n",
       "timecourse_AAACCGTACCCGCTGT               0.084507  ...         48   \n",
       "timecourse_GTAGGTTACCCGCTGT               0.239437  ...         96   \n",
       "timecourse_CTAGTAAACCCGCTGT               0.183099  ...         34   \n",
       "timecourse_TGGCATGACCCGCTGT               0.154930  ...         47   \n",
       "timecourse_TTTGTGCACCCGCTGT               0.140845  ...         11   \n",
       "...                                            ...  ...        ...   \n",
       "cocultured28_AGCAATTGACACCTAC                  NaN  ...         22   \n",
       "cocultured28_TGATTAGACACCAAAC                  NaN  ...         88   \n",
       "cocultured28_GTTAAAGCTTGAGAAC                  NaN  ...         15   \n",
       "cocultured28_ACTAAGACTTGAGAAC                  NaN  ...         33   \n",
       "cocultured28_AAGGAAGCTTGAGAAC                  NaN  ...         33   \n",
       "\n",
       "                              wsnn_res.10 wsnn_res.12  wsnn_res.14  \\\n",
       "timecourse_AAACCGTACCCGCTGT            44          31           27   \n",
       "timecourse_GTAGGTTACCCGCTGT            96         109          114   \n",
       "timecourse_CTAGTAAACCCGCTGT            32          26           23   \n",
       "timecourse_TGGCATGACCCGCTGT            49          50           47   \n",
       "timecourse_TTTGTGCACCCGCTGT            12          41           36   \n",
       "...                                   ...         ...          ...   \n",
       "cocultured28_AGCAATTGACACCTAC          14          55           49   \n",
       "cocultured28_TGATTAGACACCAAAC          89         100          103   \n",
       "cocultured28_GTTAAAGCTTGAGAAC          70          79           78   \n",
       "cocultured28_ACTAAGACTTGAGAAC          61          29           21   \n",
       "cocultured28_AAGGAAGCTTGAGAAC          43          29           21   \n",
       "\n",
       "                               wsnn_res.16  wsnn_res.18  wsnn_res.20  \\\n",
       "timecourse_AAACCGTACCCGCTGT             51           35           25   \n",
       "timecourse_GTAGGTTACCCGCTGT            120          115          117   \n",
       "timecourse_CTAGTAAACCCGCTGT             14           62           50   \n",
       "timecourse_TGGCATGACCCGCTGT             34          147          151   \n",
       "timecourse_TTTGTGCACCCGCTGT             70           52           51   \n",
       "...                                    ...          ...          ...   \n",
       "cocultured28_AGCAATTGACACCTAC          107          105           72   \n",
       "cocultured28_TGATTAGACACCAAAC          105           93           85   \n",
       "cocultured28_GTTAAAGCTTGAGAAC           73           50           39   \n",
       "cocultured28_ACTAAGACTTGAGAAC           45           73          123   \n",
       "cocultured28_AAGGAAGCTTGAGAAC           45          134           48   \n",
       "\n",
       "                               celltype_wnn           barcode  \\\n",
       "timecourse_AAACCGTACCCGCTGT        diff-NPC  AAACCGTACCCGCTGT   \n",
       "timecourse_GTAGGTTACCCGCTGT        neuron-1  GTAGGTTACCCGCTGT   \n",
       "timecourse_CTAGTAAACCCGCTGT     diff-neuron  CTAGTAAACCCGCTGT   \n",
       "timecourse_TGGCATGACCCGCTGT         hiPSC-2  TGGCATGACCCGCTGT   \n",
       "timecourse_TTTGTGCACCCGCTGT        diff-NPC  TTTGTGCACCCGCTGT   \n",
       "...                                     ...               ...   \n",
       "cocultured28_AGCAATTGACACCTAC      neuron-2  AGCAATTGACACCTAC   \n",
       "cocultured28_TGATTAGACACCAAAC      neuron-4  TGATTAGACACCAAAC   \n",
       "cocultured28_GTTAAAGCTTGAGAAC      neuron-3  GTTAAAGCTTGAGAAC   \n",
       "cocultured28_ACTAAGACTTGAGAAC      neuron-3  ACTAAGACTTGAGAAC   \n",
       "cocultured28_AAGGAAGCTTGAGAAC      neuron-3  AAGGAAGCTTGAGAAC   \n",
       "\n",
       "                                                long_barcode  \n",
       "timecourse_AAACCGTACCCGCTGT      timecourse_AAACCGTACCCGCTGT  \n",
       "timecourse_GTAGGTTACCCGCTGT      timecourse_GTAGGTTACCCGCTGT  \n",
       "timecourse_CTAGTAAACCCGCTGT      timecourse_CTAGTAAACCCGCTGT  \n",
       "timecourse_TGGCATGACCCGCTGT      timecourse_TGGCATGACCCGCTGT  \n",
       "timecourse_TTTGTGCACCCGCTGT      timecourse_TTTGTGCACCCGCTGT  \n",
       "...                                                      ...  \n",
       "cocultured28_AGCAATTGACACCTAC  cocultured28_AGCAATTGACACCTAC  \n",
       "cocultured28_TGATTAGACACCAAAC  cocultured28_TGATTAGACACCAAAC  \n",
       "cocultured28_GTTAAAGCTTGAGAAC  cocultured28_GTTAAAGCTTGAGAAC  \n",
       "cocultured28_ACTAAGACTTGAGAAC  cocultured28_ACTAAGACTTGAGAAC  \n",
       "cocultured28_AAGGAAGCTTGAGAAC  cocultured28_AAGGAAGCTTGAGAAC  \n",
       "\n",
       "[22367 rows x 55 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a9adc5ad-38c1-4148-b572-5a745b615a5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(adata.obs_names)) == len(adata.obs_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9d5b09d3-6b76-4c2f-b27c-8ba77ce6ca9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# do the same for the cistopic object\n",
    "cistopic_obj.cell_data[\"long_barcode\"] = cistopic_obj.cell_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e1292c1a-c1bb-4082-99c7-03287d0a65ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19622"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# check if there are common barcodes\n",
    "len(list(set(adata.obs_names.copy(deep=True)) & set(list(cistopic_obj.cell_names.copy())))) #128\n",
    "#len(list(set(adata.obs_names.copy(deep=True)))) #757\n",
    "#len(set(list(cistopic_obj.cell_names.copy())))  #128\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2b1fc7e3-4a10-4928-bcf2-ee02cafa74d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['timecourse_TCAATTGCTCCTGAGT', 'timecourse_TGAACCGACTATCCTG',\n",
       "       'timecourse_GGATTAGTGGCTTTAA', ...,\n",
       "       'cocultured28_CCTAAGCCTTGAGGTA', 'cocultured28_TCCACAATGCGTGATT',\n",
       "       'cocultured28_GTGCGAGTGCAAGCTT'], dtype=object)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cistopic_obj.selected_model.cell_topic.columns.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "23709bb3-50a6-49f0-9d52-8174c3544c34",
   "metadata": {},
   "outputs": [],
   "source": [
    "#from pycisTopic.cistopic_class import *\n",
    "#from pycisTopic.diff_features import *\n",
    "#common_cells = list(set(adata.obs_names.copy(deep=True)) & set(list(cistopic_obj.cell_names.copy())))\n",
    "#impute_accessibility(cistopic_obj, selected_cells=common_cells)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4a9b086b-edc7-4b64-b63f-945c4dc5d5fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "22367\n",
      "22367\n",
      "True\n",
      "19827\n",
      "19827\n"
     ]
    }
   ],
   "source": [
    "print(len(adata.obs_names.copy(deep=True)) == len(set(adata.obs_names.copy(deep=True))))\n",
    "print(len(adata.obs_names.copy(deep=True).drop_duplicates(keep='first')))\n",
    "print(len(set(adata.obs_names.copy(deep=True))))\n",
    "print(len(list(cistopic_obj.cell_names.copy())) == len(set(list(cistopic_obj.cell_names.copy()))))\n",
    "print(len(list(cistopic_obj.cell_names.copy())))\n",
    "print(len(set(list(cistopic_obj.cell_names.copy()))))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cbebe8f",
   "metadata": {},
   "source": [
    "## Create SCENIC+ object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f1ac4cc0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-05-28 10:48:23,288 cisTopic     INFO     Imputing drop-outs\n",
      "2023-05-28 10:48:56,811 cisTopic     INFO     Scaling\n",
      "2023-05-28 10:49:36,280 cisTopic     INFO     Keep non zero rows\n",
      "2023-05-28 10:50:13,752 cisTopic     INFO     Imputed accessibility sparsity: 0.6699118462550895\n",
      "2023-05-28 10:50:13,754 cisTopic     INFO     Create CistopicImputedFeatures object\n",
      "2023-05-28 10:50:13,755 cisTopic     INFO     Done!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SCENIC+ object with n_cells x n_genes = 19622 x 30768 and n_cells x n_regions = 19622 x 538878\n",
       "\tmetadata_regions:'Chromosome', 'Start', 'End', 'Width', 'cisTopic_nr_frag', 'cisTopic_log_nr_frag', 'cisTopic_nr_acc', 'cisTopic_log_nr_acc'\n",
       "\tmetadata_genes:'features'\n",
       "\tmetadata_cell:'GEX_orig.ident', 'GEX_nCount_RNA', 'GEX_nFeature_RNA', 'GEX_percent.mt', 'GEX_percent.ribo', 'GEX_nCount_SCT', 'GEX_nFeature_SCT', 'GEX_SCT_snn_res.0.5', 'GEX_seurat_clusters', 'GEX_pANN_0.25_0.005_794', 'GEX_DF.classifications_0.25_0.005_794', 'GEX_doubletClass', 'GEX_bc', 'GEX_nCount_ATAC', 'GEX_nFeature_ATAC', 'GEX_SCT.weight', 'GEX_ATAC.weight', 'GEX_pANN_0.25_0.005_398', 'GEX_DF.classifications_0.25_0.005_398', 'GEX_pANN_0.25_0.005_1306', 'GEX_DF.classifications_0.25_0.005_1306', 'GEX_pANN_0.25_0.005_395', 'GEX_DF.classifications_0.25_0.005_395', 'GEX_ident', 'GEX_scDblFinder.sample', 'GEX_scDblFinder.class', 'GEX_scDblFinder.score', 'GEX_scDblFinder.weighted', 'GEX_scDblFinder.cxds_score', 'GEX_res.0.5', 'GEX_celltype', 'GEX_pseudotime', 'GEX_wsnn_res.0.8', 'GEX_wsnn_res.0.1', 'GEX_wsnn_res.0.25', 'GEX_wsnn_res.0.5', 'GEX_wsnn_res.0.75', 'GEX_wsnn_res.1', 'GEX_wsnn_res.2', 'GEX_wsnn_res.3', 'GEX_wsnn_res.4', 'GEX_wsnn_res.5', 'GEX_wsnn_res.6', 'GEX_wsnn_res.7', 'GEX_wsnn_res.8', 'GEX_wsnn_res.9', 'GEX_wsnn_res.10', 'GEX_wsnn_res.12', 'GEX_wsnn_res.14', 'GEX_wsnn_res.16', 'GEX_wsnn_res.18', 'GEX_wsnn_res.20', 'GEX_celltype_wnn', 'GEX_barcode', 'GEX_long_barcode', 'ACC_Dupl_rate', 'ACC_cisTopic_log_nr_acc', 'ACC_Total_nr_frag_in_regions', 'ACC_Log_total_nr_frag', 'ACC_Total_nr_frag', 'ACC_cisTopic_nr_frag', 'ACC_Unique_nr_frag_in_regions', 'ACC_Log_unique_nr_frag', 'ACC_cisTopic_nr_acc', 'ACC_cisTopic_log_nr_frag', 'ACC_Unique_nr_frag', 'ACC_Dupl_nr_frag', 'ACC_TSS_enrichment', 'ACC_FRIP', 'ACC_orig.ident', 'ACC_nCount_RNA', 'ACC_nFeature_RNA', 'ACC_percent.mt', 'ACC_percent.ribo', 'ACC_nCount_SCT', 'ACC_nFeature_SCT', 'ACC_SCT_snn_res.0.5', 'ACC_seurat_clusters', 'ACC_pANN_0.25_0.005_794', 'ACC_DF.classifications_0.25_0.005_794', 'ACC_doubletClass', 'ACC_bc', 'ACC_nCount_ATAC', 'ACC_nFeature_ATAC', 'ACC_SCT.weight', 'ACC_ATAC.weight', 'ACC_pANN_0.25_0.005_398', 'ACC_DF.classifications_0.25_0.005_398', 'ACC_pANN_0.25_0.005_1306', 'ACC_DF.classifications_0.25_0.005_1306', 'ACC_pANN_0.25_0.005_395', 'ACC_DF.classifications_0.25_0.005_395', 'ACC_ident', 'ACC_scDblFinder.sample', 'ACC_scDblFinder.class', 'ACC_scDblFinder.score', 'ACC_scDblFinder.weighted', 'ACC_scDblFinder.cxds_score', 'ACC_res.0.5', 'ACC_celltype', 'ACC_pseudotime', 'ACC_wsnn_res.0.8', 'ACC_wsnn_res.0.1', 'ACC_wsnn_res.0.25', 'ACC_wsnn_res.0.5', 'ACC_wsnn_res.0.75', 'ACC_wsnn_res.1', 'ACC_wsnn_res.2', 'ACC_wsnn_res.3', 'ACC_wsnn_res.4', 'ACC_wsnn_res.5', 'ACC_wsnn_res.6', 'ACC_wsnn_res.7', 'ACC_wsnn_res.8', 'ACC_wsnn_res.9', 'ACC_wsnn_res.10', 'ACC_wsnn_res.12', 'ACC_wsnn_res.14', 'ACC_wsnn_res.16', 'ACC_wsnn_res.18', 'ACC_wsnn_res.20', 'ACC_celltype_wnn', 'ACC_barcode', 'ACC_sample_id', 'ACC_long_barcode'\n",
       "\tmenr:'CTX_topics_otsu_All', 'CTX_topics_otsu_No_promoters', 'DEM_topics_otsu_All', 'DEM_topics_otsu_No_promoters', 'CTX_topics_top_3_All', 'CTX_topics_top_3_No_promoters', 'DEM_topics_top_3_All', 'DEM_topics_top_3_No_promoters', 'CTX_DARs_All', 'CTX_DARs_No_promoters', 'DEM_DARs_All', 'DEM_DARs_No_promoters'\n",
       "\tdr_cell:'GEX_X_umap', 'GEX_X_umap.atac', 'GEX_X_wnn.umap'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create the Scenic+ object\n",
    "from scenicplus.scenicplus_class import create_SCENICPLUS_object\n",
    "import numpy as np\n",
    "scplus_obj = create_SCENICPLUS_object(\n",
    "    GEX_anndata = adata,\n",
    "    cisTopic_obj = cistopic_obj,\n",
    "    menr = menr,\n",
    "    gene_metadata = adata.var.copy(deep=True),\n",
    "    bc_transform_func = None, #lambda x: f'{x}_timecourse' #None, #function to convert scATAC-seq barcodes to scRNA-seq ones\n",
    ")\n",
    "scplus_obj.X_EXP = np.array(scplus_obj.X_EXP.todense())\n",
    "scplus_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6691456b-a9e3-4aeb-96fd-b59ad8e31c8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#scplus_obj.add_gene_data(adata.var.copy(deep=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "341ed9d0-1d80-4aaf-b38d-22e74e1bfd9a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['MIR1302-2HG', 'AL627309.1', 'AL627309.5', 'AL627309.4', 'AP006222.2',\n",
       "       'AC114498.1', 'AL669831.2', 'LINC01409', 'FAM87B', 'LINC01128',\n",
       "       ...\n",
       "       'HSFX3', 'HSFX4', 'MAGEA4', 'U82671.1', 'AVPR2', 'HCFC1-AS1',\n",
       "       'AC009494.2', 'AC136616.3', 'AC023491.2', 'AC007325.1'],\n",
       "      dtype='object', length=30768)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj.gene_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4aa1fab7-d5ac-445b-879d-b4bd1ac153f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['MIR1302-2HG', 'AL627309.1', 'AL627309.5', 'AL627309.4', 'AP006222.2',\n",
       "       'AC114498.1', 'AL669831.2', 'LINC01409', 'FAM87B', 'LINC01128',\n",
       "       ...\n",
       "       'HSFX3', 'HSFX4', 'MAGEA4', 'U82671.1', 'AVPR2', 'HCFC1-AS1',\n",
       "       'AC009494.2', 'AC136616.3', 'AC023491.2', 'AC007325.1'],\n",
       "      dtype='object', length=30768)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var\n",
    "adata.var.copy(deep=True)\n",
    "adata.var.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "21964f78-99e8-4bf1-9bbb-d14983707b46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>percent.ribo</th>\n",
       "      <th>nCount_SCT</th>\n",
       "      <th>nFeature_SCT</th>\n",
       "      <th>SCT_snn_res.0.5</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>pANN_0.25_0.005_794</th>\n",
       "      <th>...</th>\n",
       "      <th>wsnn_res.9</th>\n",
       "      <th>wsnn_res.10</th>\n",
       "      <th>wsnn_res.12</th>\n",
       "      <th>wsnn_res.14</th>\n",
       "      <th>wsnn_res.16</th>\n",
       "      <th>wsnn_res.18</th>\n",
       "      <th>wsnn_res.20</th>\n",
       "      <th>celltype_wnn</th>\n",
       "      <th>barcode</th>\n",
       "      <th>long_barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>timecourse_AAACCGTACCCGCTGT</th>\n",
       "      <td>timecourse</td>\n",
       "      <td>5020.0</td>\n",
       "      <td>2676</td>\n",
       "      <td>6.155378</td>\n",
       "      <td>3.705179</td>\n",
       "      <td>6037.0</td>\n",
       "      <td>2676</td>\n",
       "      <td>7</td>\n",
       "      <td>25</td>\n",
       "      <td>0.084507</td>\n",
       "      <td>...</td>\n",
       "      <td>48</td>\n",
       "      <td>44</td>\n",
       "      <td>31</td>\n",
       "      <td>27</td>\n",
       "      <td>51</td>\n",
       "      <td>35</td>\n",
       "      <td>25</td>\n",
       "      <td>diff-NPC</td>\n",
       "      <td>AAACCGTACCCGCTGT</td>\n",
       "      <td>timecourse_AAACCGTACCCGCTGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>timecourse_GTAGGTTACCCGCTGT</th>\n",
       "      <td>timecourse</td>\n",
       "      <td>4719.0</td>\n",
       "      <td>2179</td>\n",
       "      <td>0.190718</td>\n",
       "      <td>0.466200</td>\n",
       "      <td>6032.0</td>\n",
       "      <td>2178</td>\n",
       "      <td>14</td>\n",
       "      <td>117</td>\n",
       "      <td>0.239437</td>\n",
       "      <td>...</td>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "      <td>109</td>\n",
       "      <td>114</td>\n",
       "      <td>120</td>\n",
       "      <td>115</td>\n",
       "      <td>117</td>\n",
       "      <td>neuron-1</td>\n",
       "      <td>GTAGGTTACCCGCTGT</td>\n",
       "      <td>timecourse_GTAGGTTACCCGCTGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>timecourse_CTAGTAAACCCGCTGT</th>\n",
       "      <td>timecourse</td>\n",
       "      <td>3832.0</td>\n",
       "      <td>2004</td>\n",
       "      <td>0.287056</td>\n",
       "      <td>0.730689</td>\n",
       "      <td>5979.0</td>\n",
       "      <td>2008</td>\n",
       "      <td>12</td>\n",
       "      <td>50</td>\n",
       "      <td>0.183099</td>\n",
       "      <td>...</td>\n",
       "      <td>34</td>\n",
       "      <td>32</td>\n",
       "      <td>26</td>\n",
       "      <td>23</td>\n",
       "      <td>14</td>\n",
       "      <td>62</td>\n",
       "      <td>50</td>\n",
       "      <td>diff-neuron</td>\n",
       "      <td>CTAGTAAACCCGCTGT</td>\n",
       "      <td>timecourse_CTAGTAAACCCGCTGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>timecourse_TGGCATGACCCGCTGT</th>\n",
       "      <td>timecourse</td>\n",
       "      <td>7329.0</td>\n",
       "      <td>3535</td>\n",
       "      <td>1.159776</td>\n",
       "      <td>1.132487</td>\n",
       "      <td>7247.0</td>\n",
       "      <td>3535</td>\n",
       "      <td>8</td>\n",
       "      <td>151</td>\n",
       "      <td>0.154930</td>\n",
       "      <td>...</td>\n",
       "      <td>47</td>\n",
       "      <td>49</td>\n",
       "      <td>50</td>\n",
       "      <td>47</td>\n",
       "      <td>34</td>\n",
       "      <td>147</td>\n",
       "      <td>151</td>\n",
       "      <td>hiPSC-2</td>\n",
       "      <td>TGGCATGACCCGCTGT</td>\n",
       "      <td>timecourse_TGGCATGACCCGCTGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>timecourse_TTTGTGCACCCGCTGT</th>\n",
       "      <td>timecourse</td>\n",
       "      <td>4511.0</td>\n",
       "      <td>2253</td>\n",
       "      <td>7.337619</td>\n",
       "      <td>1.906451</td>\n",
       "      <td>6061.0</td>\n",
       "      <td>2253</td>\n",
       "      <td>7</td>\n",
       "      <td>51</td>\n",
       "      <td>0.140845</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>41</td>\n",
       "      <td>36</td>\n",
       "      <td>70</td>\n",
       "      <td>52</td>\n",
       "      <td>51</td>\n",
       "      <td>diff-NPC</td>\n",
       "      <td>TTTGTGCACCCGCTGT</td>\n",
       "      <td>timecourse_TTTGTGCACCCGCTGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cocultured28_AGCAATTGACACCTAC</th>\n",
       "      <td>cocultured28</td>\n",
       "      <td>11717.0</td>\n",
       "      <td>3893</td>\n",
       "      <td>0.162158</td>\n",
       "      <td>0.187761</td>\n",
       "      <td>7918.0</td>\n",
       "      <td>3746</td>\n",
       "      <td>5</td>\n",
       "      <td>72</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>22</td>\n",
       "      <td>14</td>\n",
       "      <td>55</td>\n",
       "      <td>49</td>\n",
       "      <td>107</td>\n",
       "      <td>105</td>\n",
       "      <td>72</td>\n",
       "      <td>neuron-2</td>\n",
       "      <td>AGCAATTGACACCTAC</td>\n",
       "      <td>cocultured28_AGCAATTGACACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cocultured28_TGATTAGACACCAAAC</th>\n",
       "      <td>cocultured28</td>\n",
       "      <td>6845.0</td>\n",
       "      <td>2998</td>\n",
       "      <td>0.613587</td>\n",
       "      <td>0.555150</td>\n",
       "      <td>6845.0</td>\n",
       "      <td>2998</td>\n",
       "      <td>14</td>\n",
       "      <td>85</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>88</td>\n",
       "      <td>89</td>\n",
       "      <td>100</td>\n",
       "      <td>103</td>\n",
       "      <td>105</td>\n",
       "      <td>93</td>\n",
       "      <td>85</td>\n",
       "      <td>neuron-4</td>\n",
       "      <td>TGATTAGACACCAAAC</td>\n",
       "      <td>cocultured28_TGATTAGACACCAAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cocultured28_GTTAAAGCTTGAGAAC</th>\n",
       "      <td>cocultured28</td>\n",
       "      <td>9557.0</td>\n",
       "      <td>3435</td>\n",
       "      <td>0.292979</td>\n",
       "      <td>0.282515</td>\n",
       "      <td>7697.0</td>\n",
       "      <td>3433</td>\n",
       "      <td>6</td>\n",
       "      <td>39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>70</td>\n",
       "      <td>79</td>\n",
       "      <td>78</td>\n",
       "      <td>73</td>\n",
       "      <td>50</td>\n",
       "      <td>39</td>\n",
       "      <td>neuron-3</td>\n",
       "      <td>GTTAAAGCTTGAGAAC</td>\n",
       "      <td>cocultured28_GTTAAAGCTTGAGAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cocultured28_ACTAAGACTTGAGAAC</th>\n",
       "      <td>cocultured28</td>\n",
       "      <td>1732.0</td>\n",
       "      <td>1112</td>\n",
       "      <td>2.078522</td>\n",
       "      <td>0.577367</td>\n",
       "      <td>5364.0</td>\n",
       "      <td>1614</td>\n",
       "      <td>6</td>\n",
       "      <td>123</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>33</td>\n",
       "      <td>61</td>\n",
       "      <td>29</td>\n",
       "      <td>21</td>\n",
       "      <td>45</td>\n",
       "      <td>73</td>\n",
       "      <td>123</td>\n",
       "      <td>neuron-3</td>\n",
       "      <td>ACTAAGACTTGAGAAC</td>\n",
       "      <td>cocultured28_ACTAAGACTTGAGAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cocultured28_AAGGAAGCTTGAGAAC</th>\n",
       "      <td>cocultured28</td>\n",
       "      <td>8505.0</td>\n",
       "      <td>3170</td>\n",
       "      <td>0.282187</td>\n",
       "      <td>0.435038</td>\n",
       "      <td>7565.0</td>\n",
       "      <td>3170</td>\n",
       "      <td>6</td>\n",
       "      <td>48</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>33</td>\n",
       "      <td>43</td>\n",
       "      <td>29</td>\n",
       "      <td>21</td>\n",
       "      <td>45</td>\n",
       "      <td>134</td>\n",
       "      <td>48</td>\n",
       "      <td>neuron-3</td>\n",
       "      <td>AAGGAAGCTTGAGAAC</td>\n",
       "      <td>cocultured28_AAGGAAGCTTGAGAAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22367 rows × 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 orig.ident  nCount_RNA  nFeature_RNA  \\\n",
       "timecourse_AAACCGTACCCGCTGT      timecourse      5020.0          2676   \n",
       "timecourse_GTAGGTTACCCGCTGT      timecourse      4719.0          2179   \n",
       "timecourse_CTAGTAAACCCGCTGT      timecourse      3832.0          2004   \n",
       "timecourse_TGGCATGACCCGCTGT      timecourse      7329.0          3535   \n",
       "timecourse_TTTGTGCACCCGCTGT      timecourse      4511.0          2253   \n",
       "...                                     ...         ...           ...   \n",
       "cocultured28_AGCAATTGACACCTAC  cocultured28     11717.0          3893   \n",
       "cocultured28_TGATTAGACACCAAAC  cocultured28      6845.0          2998   \n",
       "cocultured28_GTTAAAGCTTGAGAAC  cocultured28      9557.0          3435   \n",
       "cocultured28_ACTAAGACTTGAGAAC  cocultured28      1732.0          1112   \n",
       "cocultured28_AAGGAAGCTTGAGAAC  cocultured28      8505.0          3170   \n",
       "\n",
       "                               percent.mt  percent.ribo  nCount_SCT  \\\n",
       "timecourse_AAACCGTACCCGCTGT      6.155378      3.705179      6037.0   \n",
       "timecourse_GTAGGTTACCCGCTGT      0.190718      0.466200      6032.0   \n",
       "timecourse_CTAGTAAACCCGCTGT      0.287056      0.730689      5979.0   \n",
       "timecourse_TGGCATGACCCGCTGT      1.159776      1.132487      7247.0   \n",
       "timecourse_TTTGTGCACCCGCTGT      7.337619      1.906451      6061.0   \n",
       "...                                   ...           ...         ...   \n",
       "cocultured28_AGCAATTGACACCTAC    0.162158      0.187761      7918.0   \n",
       "cocultured28_TGATTAGACACCAAAC    0.613587      0.555150      6845.0   \n",
       "cocultured28_GTTAAAGCTTGAGAAC    0.292979      0.282515      7697.0   \n",
       "cocultured28_ACTAAGACTTGAGAAC    2.078522      0.577367      5364.0   \n",
       "cocultured28_AAGGAAGCTTGAGAAC    0.282187      0.435038      7565.0   \n",
       "\n",
       "                               nFeature_SCT SCT_snn_res.0.5 seurat_clusters  \\\n",
       "timecourse_AAACCGTACCCGCTGT            2676               7              25   \n",
       "timecourse_GTAGGTTACCCGCTGT            2178              14             117   \n",
       "timecourse_CTAGTAAACCCGCTGT            2008              12              50   \n",
       "timecourse_TGGCATGACCCGCTGT            3535               8             151   \n",
       "timecourse_TTTGTGCACCCGCTGT            2253               7              51   \n",
       "...                                     ...             ...             ...   \n",
       "cocultured28_AGCAATTGACACCTAC          3746               5              72   \n",
       "cocultured28_TGATTAGACACCAAAC          2998              14              85   \n",
       "cocultured28_GTTAAAGCTTGAGAAC          3433               6              39   \n",
       "cocultured28_ACTAAGACTTGAGAAC          1614               6             123   \n",
       "cocultured28_AAGGAAGCTTGAGAAC          3170               6              48   \n",
       "\n",
       "                               pANN_0.25_0.005_794  ... wsnn_res.9  \\\n",
       "timecourse_AAACCGTACCCGCTGT               0.084507  ...         48   \n",
       "timecourse_GTAGGTTACCCGCTGT               0.239437  ...         96   \n",
       "timecourse_CTAGTAAACCCGCTGT               0.183099  ...         34   \n",
       "timecourse_TGGCATGACCCGCTGT               0.154930  ...         47   \n",
       "timecourse_TTTGTGCACCCGCTGT               0.140845  ...         11   \n",
       "...                                            ...  ...        ...   \n",
       "cocultured28_AGCAATTGACACCTAC                  NaN  ...         22   \n",
       "cocultured28_TGATTAGACACCAAAC                  NaN  ...         88   \n",
       "cocultured28_GTTAAAGCTTGAGAAC                  NaN  ...         15   \n",
       "cocultured28_ACTAAGACTTGAGAAC                  NaN  ...         33   \n",
       "cocultured28_AAGGAAGCTTGAGAAC                  NaN  ...         33   \n",
       "\n",
       "                              wsnn_res.10 wsnn_res.12  wsnn_res.14  \\\n",
       "timecourse_AAACCGTACCCGCTGT            44          31           27   \n",
       "timecourse_GTAGGTTACCCGCTGT            96         109          114   \n",
       "timecourse_CTAGTAAACCCGCTGT            32          26           23   \n",
       "timecourse_TGGCATGACCCGCTGT            49          50           47   \n",
       "timecourse_TTTGTGCACCCGCTGT            12          41           36   \n",
       "...                                   ...         ...          ...   \n",
       "cocultured28_AGCAATTGACACCTAC          14          55           49   \n",
       "cocultured28_TGATTAGACACCAAAC          89         100          103   \n",
       "cocultured28_GTTAAAGCTTGAGAAC          70          79           78   \n",
       "cocultured28_ACTAAGACTTGAGAAC          61          29           21   \n",
       "cocultured28_AAGGAAGCTTGAGAAC          43          29           21   \n",
       "\n",
       "                               wsnn_res.16  wsnn_res.18  wsnn_res.20  \\\n",
       "timecourse_AAACCGTACCCGCTGT             51           35           25   \n",
       "timecourse_GTAGGTTACCCGCTGT            120          115          117   \n",
       "timecourse_CTAGTAAACCCGCTGT             14           62           50   \n",
       "timecourse_TGGCATGACCCGCTGT             34          147          151   \n",
       "timecourse_TTTGTGCACCCGCTGT             70           52           51   \n",
       "...                                    ...          ...          ...   \n",
       "cocultured28_AGCAATTGACACCTAC          107          105           72   \n",
       "cocultured28_TGATTAGACACCAAAC          105           93           85   \n",
       "cocultured28_GTTAAAGCTTGAGAAC           73           50           39   \n",
       "cocultured28_ACTAAGACTTGAGAAC           45           73          123   \n",
       "cocultured28_AAGGAAGCTTGAGAAC           45          134           48   \n",
       "\n",
       "                               celltype_wnn           barcode  \\\n",
       "timecourse_AAACCGTACCCGCTGT        diff-NPC  AAACCGTACCCGCTGT   \n",
       "timecourse_GTAGGTTACCCGCTGT        neuron-1  GTAGGTTACCCGCTGT   \n",
       "timecourse_CTAGTAAACCCGCTGT     diff-neuron  CTAGTAAACCCGCTGT   \n",
       "timecourse_TGGCATGACCCGCTGT         hiPSC-2  TGGCATGACCCGCTGT   \n",
       "timecourse_TTTGTGCACCCGCTGT        diff-NPC  TTTGTGCACCCGCTGT   \n",
       "...                                     ...               ...   \n",
       "cocultured28_AGCAATTGACACCTAC      neuron-2  AGCAATTGACACCTAC   \n",
       "cocultured28_TGATTAGACACCAAAC      neuron-4  TGATTAGACACCAAAC   \n",
       "cocultured28_GTTAAAGCTTGAGAAC      neuron-3  GTTAAAGCTTGAGAAC   \n",
       "cocultured28_ACTAAGACTTGAGAAC      neuron-3  ACTAAGACTTGAGAAC   \n",
       "cocultured28_AAGGAAGCTTGAGAAC      neuron-3  AAGGAAGCTTGAGAAC   \n",
       "\n",
       "                                                long_barcode  \n",
       "timecourse_AAACCGTACCCGCTGT      timecourse_AAACCGTACCCGCTGT  \n",
       "timecourse_GTAGGTTACCCGCTGT      timecourse_GTAGGTTACCCGCTGT  \n",
       "timecourse_CTAGTAAACCCGCTGT      timecourse_CTAGTAAACCCGCTGT  \n",
       "timecourse_TGGCATGACCCGCTGT      timecourse_TGGCATGACCCGCTGT  \n",
       "timecourse_TTTGTGCACCCGCTGT      timecourse_TTTGTGCACCCGCTGT  \n",
       "...                                                      ...  \n",
       "cocultured28_AGCAATTGACACCTAC  cocultured28_AGCAATTGACACCTAC  \n",
       "cocultured28_TGATTAGACACCAAAC  cocultured28_TGATTAGACACCAAAC  \n",
       "cocultured28_GTTAAAGCTTGAGAAC  cocultured28_GTTAAAGCTTGAGAAC  \n",
       "cocultured28_ACTAAGACTTGAGAAC  cocultured28_ACTAAGACTTGAGAAC  \n",
       "cocultured28_AAGGAAGCTTGAGAAC  cocultured28_AAGGAAGCTTGAGAAC  \n",
       "\n",
       "[22367 rows x 55 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "7293d380-3f0f-4c1f-9f19-02dc7f4bc64a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dupl_rate</th>\n",
       "      <th>cisTopic_log_nr_acc</th>\n",
       "      <th>Total_nr_frag_in_regions</th>\n",
       "      <th>Log_total_nr_frag</th>\n",
       "      <th>Total_nr_frag</th>\n",
       "      <th>cisTopic_nr_frag</th>\n",
       "      <th>Unique_nr_frag_in_regions</th>\n",
       "      <th>Log_unique_nr_frag</th>\n",
       "      <th>cisTopic_nr_acc</th>\n",
       "      <th>cisTopic_log_nr_frag</th>\n",
       "      <th>...</th>\n",
       "      <th>wsnn_res.10</th>\n",
       "      <th>wsnn_res.12</th>\n",
       "      <th>wsnn_res.14</th>\n",
       "      <th>wsnn_res.16</th>\n",
       "      <th>wsnn_res.18</th>\n",
       "      <th>wsnn_res.20</th>\n",
       "      <th>celltype_wnn</th>\n",
       "      <th>barcode</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>long_barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>timecourse_TCAATTGCTCCTGAGT</th>\n",
       "      <td>0.992421</td>\n",
       "      <td>3.625621</td>\n",
       "      <td>470578</td>\n",
       "      <td>5.885086</td>\n",
       "      <td>767513</td>\n",
       "      <td>4379</td>\n",
       "      <td>3350</td>\n",
       "      <td>3.764699</td>\n",
       "      <td>4223</td>\n",
       "      <td>3.641375</td>\n",
       "      <td>...</td>\n",
       "      <td>49</td>\n",
       "      <td>50</td>\n",
       "      <td>47</td>\n",
       "      <td>34</td>\n",
       "      <td>147</td>\n",
       "      <td>151</td>\n",
       "      <td>hiPSC-2</td>\n",
       "      <td>TCAATTGCTCCTGAGT</td>\n",
       "      <td>timecourse</td>\n",
       "      <td>timecourse_TCAATTGCTCCTGAGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>timecourse_TGAACCGACTATCCTG</th>\n",
       "      <td>0.992839</td>\n",
       "      <td>3.804957</td>\n",
       "      <td>781384</td>\n",
       "      <td>6.076014</td>\n",
       "      <td>1191281</td>\n",
       "      <td>6794</td>\n",
       "      <td>5182</td>\n",
       "      <td>3.931000</td>\n",
       "      <td>6382</td>\n",
       "      <td>3.832126</td>\n",
       "      <td>...</td>\n",
       "      <td>32</td>\n",
       "      <td>26</td>\n",
       "      <td>23</td>\n",
       "      <td>14</td>\n",
       "      <td>62</td>\n",
       "      <td>50</td>\n",
       "      <td>diff-neuron</td>\n",
       "      <td>TGAACCGACTATCCTG</td>\n",
       "      <td>timecourse</td>\n",
       "      <td>timecourse_TGAACCGACTATCCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>timecourse_GGATTAGTGGCTTTAA</th>\n",
       "      <td>0.992924</td>\n",
       "      <td>4.093352</td>\n",
       "      <td>1642543</td>\n",
       "      <td>6.460567</td>\n",
       "      <td>2887798</td>\n",
       "      <td>13338</td>\n",
       "      <td>11197</td>\n",
       "      <td>4.310375</td>\n",
       "      <td>12398</td>\n",
       "      <td>4.125091</td>\n",
       "      <td>...</td>\n",
       "      <td>40</td>\n",
       "      <td>8</td>\n",
       "      <td>61</td>\n",
       "      <td>50</td>\n",
       "      <td>36</td>\n",
       "      <td>21</td>\n",
       "      <td>hiPSC-1</td>\n",
       "      <td>GGATTAGTGGCTTTAA</td>\n",
       "      <td>timecourse</td>\n",
       "      <td>timecourse_GGATTAGTGGCTTTAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>timecourse_AAGCCTAACTAATCAG</th>\n",
       "      <td>0.992291</td>\n",
       "      <td>4.273441</td>\n",
       "      <td>2496161</td>\n",
       "      <td>6.590419</td>\n",
       "      <td>3894208</td>\n",
       "      <td>20864</td>\n",
       "      <td>19777</td>\n",
       "      <td>4.477411</td>\n",
       "      <td>18769</td>\n",
       "      <td>4.319398</td>\n",
       "      <td>...</td>\n",
       "      <td>97</td>\n",
       "      <td>105</td>\n",
       "      <td>113</td>\n",
       "      <td>114</td>\n",
       "      <td>125</td>\n",
       "      <td>115</td>\n",
       "      <td>hiPSC-2</td>\n",
       "      <td>AAGCCTAACTAATCAG</td>\n",
       "      <td>timecourse</td>\n",
       "      <td>timecourse_AAGCCTAACTAATCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>timecourse_GGCGGTTACCTGGATG</th>\n",
       "      <td>0.992954</td>\n",
       "      <td>3.723948</td>\n",
       "      <td>655078</td>\n",
       "      <td>6.012438</td>\n",
       "      <td>1029054</td>\n",
       "      <td>5564</td>\n",
       "      <td>4457</td>\n",
       "      <td>3.860398</td>\n",
       "      <td>5296</td>\n",
       "      <td>3.745387</td>\n",
       "      <td>...</td>\n",
       "      <td>37</td>\n",
       "      <td>22</td>\n",
       "      <td>70</td>\n",
       "      <td>48</td>\n",
       "      <td>143</td>\n",
       "      <td>156</td>\n",
       "      <td>diff-NPC</td>\n",
       "      <td>GGCGGTTACCTGGATG</td>\n",
       "      <td>timecourse</td>\n",
       "      <td>timecourse_GGCGGTTACCTGGATG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cocultured28_GATGAAGGAGGCCCTT</th>\n",
       "      <td>0.993454</td>\n",
       "      <td>3.123525</td>\n",
       "      <td>190524</td>\n",
       "      <td>5.496168</td>\n",
       "      <td>313450</td>\n",
       "      <td>1379</td>\n",
       "      <td>1164</td>\n",
       "      <td>3.312177</td>\n",
       "      <td>1329</td>\n",
       "      <td>3.139564</td>\n",
       "      <td>...</td>\n",
       "      <td>108</td>\n",
       "      <td>127</td>\n",
       "      <td>132</td>\n",
       "      <td>148</td>\n",
       "      <td>157</td>\n",
       "      <td>164</td>\n",
       "      <td>neuron-3</td>\n",
       "      <td>GATGAAGGAGGCCCTT</td>\n",
       "      <td>cocultured28</td>\n",
       "      <td>cocultured28_GATGAAGGAGGCCCTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cocultured28_CTGTATTGATCCTAAT</th>\n",
       "      <td>0.992931</td>\n",
       "      <td>3.394802</td>\n",
       "      <td>336729</td>\n",
       "      <td>5.666258</td>\n",
       "      <td>463722</td>\n",
       "      <td>2584</td>\n",
       "      <td>2166</td>\n",
       "      <td>3.515609</td>\n",
       "      <td>2482</td>\n",
       "      <td>3.412293</td>\n",
       "      <td>...</td>\n",
       "      <td>43</td>\n",
       "      <td>29</td>\n",
       "      <td>21</td>\n",
       "      <td>45</td>\n",
       "      <td>73</td>\n",
       "      <td>48</td>\n",
       "      <td>neuron-3</td>\n",
       "      <td>CTGTATTGATCCTAAT</td>\n",
       "      <td>cocultured28</td>\n",
       "      <td>cocultured28_CTGTATTGATCCTAAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cocultured28_CCTAAGCCTTGAGGTA</th>\n",
       "      <td>0.993037</td>\n",
       "      <td>3.474216</td>\n",
       "      <td>413668</td>\n",
       "      <td>5.762289</td>\n",
       "      <td>578481</td>\n",
       "      <td>3171</td>\n",
       "      <td>2706</td>\n",
       "      <td>3.605089</td>\n",
       "      <td>2980</td>\n",
       "      <td>3.501196</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>55</td>\n",
       "      <td>49</td>\n",
       "      <td>94</td>\n",
       "      <td>80</td>\n",
       "      <td>90</td>\n",
       "      <td>neuron-2</td>\n",
       "      <td>CCTAAGCCTTGAGGTA</td>\n",
       "      <td>cocultured28</td>\n",
       "      <td>cocultured28_CCTAAGCCTTGAGGTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cocultured28_TCCACAATGCGTGATT</th>\n",
       "      <td>0.993336</td>\n",
       "      <td>3.218273</td>\n",
       "      <td>220009</td>\n",
       "      <td>5.499695</td>\n",
       "      <td>316006</td>\n",
       "      <td>1693</td>\n",
       "      <td>1345</td>\n",
       "      <td>3.323458</td>\n",
       "      <td>1653</td>\n",
       "      <td>3.228657</td>\n",
       "      <td>...</td>\n",
       "      <td>61</td>\n",
       "      <td>62</td>\n",
       "      <td>57</td>\n",
       "      <td>134</td>\n",
       "      <td>129</td>\n",
       "      <td>136</td>\n",
       "      <td>neuron-3</td>\n",
       "      <td>TCCACAATGCGTGATT</td>\n",
       "      <td>cocultured28</td>\n",
       "      <td>cocultured28_TCCACAATGCGTGATT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cocultured28_GTGCGAGTGCAAGCTT</th>\n",
       "      <td>0.992965</td>\n",
       "      <td>3.186956</td>\n",
       "      <td>196610</td>\n",
       "      <td>5.469131</td>\n",
       "      <td>294531</td>\n",
       "      <td>1590</td>\n",
       "      <td>1268</td>\n",
       "      <td>3.316390</td>\n",
       "      <td>1538</td>\n",
       "      <td>3.201397</td>\n",
       "      <td>...</td>\n",
       "      <td>61</td>\n",
       "      <td>62</td>\n",
       "      <td>57</td>\n",
       "      <td>142</td>\n",
       "      <td>152</td>\n",
       "      <td>123</td>\n",
       "      <td>neuron-3</td>\n",
       "      <td>GTGCGAGTGCAAGCTT</td>\n",
       "      <td>cocultured28</td>\n",
       "      <td>cocultured28_GTGCGAGTGCAAGCTT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19827 rows × 70 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               Dupl_rate cisTopic_log_nr_acc  \\\n",
       "timecourse_TCAATTGCTCCTGAGT     0.992421            3.625621   \n",
       "timecourse_TGAACCGACTATCCTG     0.992839            3.804957   \n",
       "timecourse_GGATTAGTGGCTTTAA     0.992924            4.093352   \n",
       "timecourse_AAGCCTAACTAATCAG     0.992291            4.273441   \n",
       "timecourse_GGCGGTTACCTGGATG     0.992954            3.723948   \n",
       "...                                  ...                 ...   \n",
       "cocultured28_GATGAAGGAGGCCCTT   0.993454            3.123525   \n",
       "cocultured28_CTGTATTGATCCTAAT   0.992931            3.394802   \n",
       "cocultured28_CCTAAGCCTTGAGGTA   0.993037            3.474216   \n",
       "cocultured28_TCCACAATGCGTGATT   0.993336            3.218273   \n",
       "cocultured28_GTGCGAGTGCAAGCTT   0.992965            3.186956   \n",
       "\n",
       "                               Total_nr_frag_in_regions  Log_total_nr_frag  \\\n",
       "timecourse_TCAATTGCTCCTGAGT                      470578           5.885086   \n",
       "timecourse_TGAACCGACTATCCTG                      781384           6.076014   \n",
       "timecourse_GGATTAGTGGCTTTAA                     1642543           6.460567   \n",
       "timecourse_AAGCCTAACTAATCAG                     2496161           6.590419   \n",
       "timecourse_GGCGGTTACCTGGATG                      655078           6.012438   \n",
       "...                                                 ...                ...   \n",
       "cocultured28_GATGAAGGAGGCCCTT                    190524           5.496168   \n",
       "cocultured28_CTGTATTGATCCTAAT                    336729           5.666258   \n",
       "cocultured28_CCTAAGCCTTGAGGTA                    413668           5.762289   \n",
       "cocultured28_TCCACAATGCGTGATT                    220009           5.499695   \n",
       "cocultured28_GTGCGAGTGCAAGCTT                    196610           5.469131   \n",
       "\n",
       "                               Total_nr_frag cisTopic_nr_frag  \\\n",
       "timecourse_TCAATTGCTCCTGAGT           767513             4379   \n",
       "timecourse_TGAACCGACTATCCTG          1191281             6794   \n",
       "timecourse_GGATTAGTGGCTTTAA          2887798            13338   \n",
       "timecourse_AAGCCTAACTAATCAG          3894208            20864   \n",
       "timecourse_GGCGGTTACCTGGATG          1029054             5564   \n",
       "...                                      ...              ...   \n",
       "cocultured28_GATGAAGGAGGCCCTT         313450             1379   \n",
       "cocultured28_CTGTATTGATCCTAAT         463722             2584   \n",
       "cocultured28_CCTAAGCCTTGAGGTA         578481             3171   \n",
       "cocultured28_TCCACAATGCGTGATT         316006             1693   \n",
       "cocultured28_GTGCGAGTGCAAGCTT         294531             1590   \n",
       "\n",
       "                               Unique_nr_frag_in_regions  Log_unique_nr_frag  \\\n",
       "timecourse_TCAATTGCTCCTGAGT                         3350            3.764699   \n",
       "timecourse_TGAACCGACTATCCTG                         5182            3.931000   \n",
       "timecourse_GGATTAGTGGCTTTAA                        11197            4.310375   \n",
       "timecourse_AAGCCTAACTAATCAG                        19777            4.477411   \n",
       "timecourse_GGCGGTTACCTGGATG                         4457            3.860398   \n",
       "...                                                  ...                 ...   \n",
       "cocultured28_GATGAAGGAGGCCCTT                       1164            3.312177   \n",
       "cocultured28_CTGTATTGATCCTAAT                       2166            3.515609   \n",
       "cocultured28_CCTAAGCCTTGAGGTA                       2706            3.605089   \n",
       "cocultured28_TCCACAATGCGTGATT                       1345            3.323458   \n",
       "cocultured28_GTGCGAGTGCAAGCTT                       1268            3.316390   \n",
       "\n",
       "                              cisTopic_nr_acc cisTopic_log_nr_frag  ...  \\\n",
       "timecourse_TCAATTGCTCCTGAGT              4223             3.641375  ...   \n",
       "timecourse_TGAACCGACTATCCTG              6382             3.832126  ...   \n",
       "timecourse_GGATTAGTGGCTTTAA             12398             4.125091  ...   \n",
       "timecourse_AAGCCTAACTAATCAG             18769             4.319398  ...   \n",
       "timecourse_GGCGGTTACCTGGATG              5296             3.745387  ...   \n",
       "...                                       ...                  ...  ...   \n",
       "cocultured28_GATGAAGGAGGCCCTT            1329             3.139564  ...   \n",
       "cocultured28_CTGTATTGATCCTAAT            2482             3.412293  ...   \n",
       "cocultured28_CCTAAGCCTTGAGGTA            2980             3.501196  ...   \n",
       "cocultured28_TCCACAATGCGTGATT            1653             3.228657  ...   \n",
       "cocultured28_GTGCGAGTGCAAGCTT            1538             3.201397  ...   \n",
       "\n",
       "                               wsnn_res.10  wsnn_res.12  wsnn_res.14  \\\n",
       "timecourse_TCAATTGCTCCTGAGT             49           50           47   \n",
       "timecourse_TGAACCGACTATCCTG             32           26           23   \n",
       "timecourse_GGATTAGTGGCTTTAA             40            8           61   \n",
       "timecourse_AAGCCTAACTAATCAG             97          105          113   \n",
       "timecourse_GGCGGTTACCTGGATG             37           22           70   \n",
       "...                                    ...          ...          ...   \n",
       "cocultured28_GATGAAGGAGGCCCTT          108          127          132   \n",
       "cocultured28_CTGTATTGATCCTAAT           43           29           21   \n",
       "cocultured28_CCTAAGCCTTGAGGTA           14           55           49   \n",
       "cocultured28_TCCACAATGCGTGATT           61           62           57   \n",
       "cocultured28_GTGCGAGTGCAAGCTT           61           62           57   \n",
       "\n",
       "                               wsnn_res.16 wsnn_res.18  wsnn_res.20  \\\n",
       "timecourse_TCAATTGCTCCTGAGT             34         147          151   \n",
       "timecourse_TGAACCGACTATCCTG             14          62           50   \n",
       "timecourse_GGATTAGTGGCTTTAA             50          36           21   \n",
       "timecourse_AAGCCTAACTAATCAG            114         125          115   \n",
       "timecourse_GGCGGTTACCTGGATG             48         143          156   \n",
       "...                                    ...         ...          ...   \n",
       "cocultured28_GATGAAGGAGGCCCTT          148         157          164   \n",
       "cocultured28_CTGTATTGATCCTAAT           45          73           48   \n",
       "cocultured28_CCTAAGCCTTGAGGTA           94          80           90   \n",
       "cocultured28_TCCACAATGCGTGATT          134         129          136   \n",
       "cocultured28_GTGCGAGTGCAAGCTT          142         152          123   \n",
       "\n",
       "                               celltype_wnn           barcode     sample_id  \\\n",
       "timecourse_TCAATTGCTCCTGAGT         hiPSC-2  TCAATTGCTCCTGAGT    timecourse   \n",
       "timecourse_TGAACCGACTATCCTG     diff-neuron  TGAACCGACTATCCTG    timecourse   \n",
       "timecourse_GGATTAGTGGCTTTAA         hiPSC-1  GGATTAGTGGCTTTAA    timecourse   \n",
       "timecourse_AAGCCTAACTAATCAG         hiPSC-2  AAGCCTAACTAATCAG    timecourse   \n",
       "timecourse_GGCGGTTACCTGGATG        diff-NPC  GGCGGTTACCTGGATG    timecourse   \n",
       "...                                     ...               ...           ...   \n",
       "cocultured28_GATGAAGGAGGCCCTT      neuron-3  GATGAAGGAGGCCCTT  cocultured28   \n",
       "cocultured28_CTGTATTGATCCTAAT      neuron-3  CTGTATTGATCCTAAT  cocultured28   \n",
       "cocultured28_CCTAAGCCTTGAGGTA      neuron-2  CCTAAGCCTTGAGGTA  cocultured28   \n",
       "cocultured28_TCCACAATGCGTGATT      neuron-3  TCCACAATGCGTGATT  cocultured28   \n",
       "cocultured28_GTGCGAGTGCAAGCTT      neuron-3  GTGCGAGTGCAAGCTT  cocultured28   \n",
       "\n",
       "                                                long_barcode  \n",
       "timecourse_TCAATTGCTCCTGAGT      timecourse_TCAATTGCTCCTGAGT  \n",
       "timecourse_TGAACCGACTATCCTG      timecourse_TGAACCGACTATCCTG  \n",
       "timecourse_GGATTAGTGGCTTTAA      timecourse_GGATTAGTGGCTTTAA  \n",
       "timecourse_AAGCCTAACTAATCAG      timecourse_AAGCCTAACTAATCAG  \n",
       "timecourse_GGCGGTTACCTGGATG      timecourse_GGCGGTTACCTGGATG  \n",
       "...                                                      ...  \n",
       "cocultured28_GATGAAGGAGGCCCTT  cocultured28_GATGAAGGAGGCCCTT  \n",
       "cocultured28_CTGTATTGATCCTAAT  cocultured28_CTGTATTGATCCTAAT  \n",
       "cocultured28_CCTAAGCCTTGAGGTA  cocultured28_CCTAAGCCTTGAGGTA  \n",
       "cocultured28_TCCACAATGCGTGATT  cocultured28_TCCACAATGCGTGATT  \n",
       "cocultured28_GTGCGAGTGCAAGCTT  cocultured28_GTGCGAGTGCAAGCTT  \n",
       "\n",
       "[19827 rows x 70 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cistopic_obj.cell_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "73e16bbe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "host: 105\n",
      "True\n",
      "Genes recovered: 20796 out of 30768\n",
      "host: 104\n",
      "True\n",
      "Genes recovered: 21035 out of 30768\n",
      "host: 103\n",
      "True\n",
      "Genes recovered: 30007 out of 30768\n",
      "host: 102\n",
      "True\n",
      "Genes recovered: 30078 out of 30768\n",
      "host: 101\n",
      "True\n",
      "Genes recovered: 30194 out of 30768\n",
      "host: 100\n",
      "True\n",
      "Genes recovered: 30378 out of 30768\n",
      "host: 99\n",
      "True\n",
      "Genes recovered: 30461 out of 30768\n",
      "host: 98\n",
      "True\n",
      "Genes recovered: 30745 out of 30768\n",
      "host: 97\n",
      "True\n",
      "Genes recovered: 30345 out of 30768\n",
      "host: 96\n",
      "True\n",
      "Genes recovered: 28859 out of 30768\n",
      "host: 95\n",
      "True\n",
      "Genes recovered: 28591 out of 30768\n",
      "host: 94\n",
      "True\n",
      "Genes recovered: 28508 out of 30768\n",
      "host: 93\n",
      "True\n",
      "Genes recovered: 28203 out of 30768\n",
      "host: 92\n",
      "True\n",
      "Genes recovered: 28094 out of 30768\n",
      "host: 91\n",
      "Host not reachable\n",
      "host: 90\n",
      "Host not reachable\n",
      "host: 89\n",
      "Host not reachable\n",
      "host: 88\n",
      "Host not reachable\n",
      "host: 87\n",
      "Host not reachable\n",
      "host: 86\n",
      "Host not reachable\n",
      "host: 80\n",
      "True\n",
      "Genes recovered: 20176 out of 30768\n",
      "host: 77\n",
      "True\n",
      "Genes recovered: 19803 out of 30768\n",
      "host: 75\n",
      "Host not reachable\n",
      "host: 54\n",
      "Host not reachable\n",
      "version: 98 has the largest overlap, use http://sep2019.archive.ensembl.org/ as biomart host\n"
     ]
    }
   ],
   "source": [
    "# Select the optimal gene names host\n",
    "ensembl_version_dict = {'105': 'http://www.ensembl.org',\n",
    "                        '104': 'http://may2021.archive.ensembl.org/',\n",
    "                        '103': 'http://feb2021.archive.ensembl.org/',\n",
    "                        '102': 'http://nov2020.archive.ensembl.org/',\n",
    "                        '101': 'http://aug2020.archive.ensembl.org/',\n",
    "                        '100': 'http://apr2020.archive.ensembl.org/',\n",
    "                        '99': 'http://jan2020.archive.ensembl.org/',\n",
    "                        '98': 'http://sep2019.archive.ensembl.org/',\n",
    "                        '97': 'http://jul2019.archive.ensembl.org/',\n",
    "                        '96': 'http://apr2019.archive.ensembl.org/',\n",
    "                        '95': 'http://jan2019.archive.ensembl.org/',\n",
    "                        '94': 'http://oct2018.archive.ensembl.org/',\n",
    "                        '93': 'http://jul2018.archive.ensembl.org/',\n",
    "                        '92': 'http://apr2018.archive.ensembl.org/',\n",
    "                        '91': 'http://dec2017.archive.ensembl.org/',\n",
    "                        '90': 'http://aug2017.archive.ensembl.org/',\n",
    "                        '89': 'http://may2017.archive.ensembl.org/',\n",
    "                        '88': 'http://mar2017.archive.ensembl.org/',\n",
    "                        '87': 'http://dec2016.archive.ensembl.org/',\n",
    "                        '86': 'http://oct2016.archive.ensembl.org/',\n",
    "                        '80': 'http://may2015.archive.ensembl.org/',\n",
    "                        '77': 'http://oct2014.archive.ensembl.org/',\n",
    "                        '75': 'http://feb2014.archive.ensembl.org/',\n",
    "                        '54': 'http://may2009.archive.ensembl.org/'}\n",
    "\n",
    "import pybiomart as pbm\n",
    "def test_ensembl_host(scplus_obj, host, species):\n",
    "    dataset = pbm.Dataset(name=species+'_gene_ensembl',  host=host)\n",
    "    annot = dataset.query(attributes=['chromosome_name', 'transcription_start_site', 'strand', 'external_gene_name', 'transcript_biotype'])\n",
    "    annot.columns = ['Chromosome', 'Start', 'Strand', 'Gene', 'Transcript_type']\n",
    "    annot['Chromosome'] = annot['Chromosome'].astype('str')\n",
    "    filter = annot['Chromosome'].str.contains('CHR|GL|JH|MT')\n",
    "    annot = annot[~filter]\n",
    "    annot.columns=['Chromosome', 'Start', 'Strand', 'Gene', 'Transcript_type']\n",
    "    gene_names_release = set(annot['Gene'].tolist())\n",
    "    #print(gene_names_release)[1:5]\n",
    "    #print(scplus_obj.gene_names)[1:5]\n",
    "    print(len(list(set(gene_names_release) & set(scplus_obj.gene_names))) > 0)\n",
    "    ov=len([x for x in scplus_obj.gene_names if x in gene_names_release])\n",
    "    print('Genes recovered: ' + str(ov) + ' out of ' + str(len(scplus_obj.gene_names)))\n",
    "    return ov\n",
    "\n",
    "n_overlap = {}\n",
    "for version in ensembl_version_dict.keys():\n",
    "    print(f'host: {version}')\n",
    "    try:\n",
    "        n_overlap[version] =  test_ensembl_host(scplus_obj, ensembl_version_dict[version], 'hsapiens')\n",
    "    except:\n",
    "        print('Host not reachable')\n",
    "v = sorted(n_overlap.items(), key=lambda item: item[1], reverse=True)[0][0]\n",
    "print(f\"version: {v} has the largest overlap, use {ensembl_version_dict[v]} as biomart host\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "44bb4a0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Choose the best host\n",
    "biomart_host = \"http://sep2019.archive.ensembl.org/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9a4b609d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-05-28 10:50:58--  http://humantfs.ccbr.utoronto.ca/download/v_1.01/TF_names_v_1.01.txt\n",
      "Resolving humantfs.ccbr.utoronto.ca (humantfs.ccbr.utoronto.ca)... 142.150.52.218\n",
      "Connecting to humantfs.ccbr.utoronto.ca (humantfs.ccbr.utoronto.ca)|142.150.52.218|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 11838 (12K) [text/plain]\n",
      "Saving to: ‘/g/scb/zaugg/deuner/SCENIC+/inputdata/utoronto_human_tfs_v_1.01.txt’\n",
      "\n",
      "/g/scb/zaugg/deuner 100%[===================>]  11.56K  --.-KB/s    in 0.1s    \n",
      "\n",
      "2023-05-28 10:50:59 (107 KB/s) - ‘/g/scb/zaugg/deuner/SCENIC+/inputdata/utoronto_human_tfs_v_1.01.txt’ saved [11838/11838]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Before running  also download a list of known human TFs from the human transcription factors database\n",
    "!wget -O /g/scb/zaugg/deuner/SCENIC+/inputdata/utoronto_human_tfs_v_1.01.txt  http://humantfs.ccbr.utoronto.ca/download/v_1.01/TF_names_v_1.01.txt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "20de72ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-05-28 10:51:01--  http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/bedToBigBed\n",
      "Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163\n",
      "Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 9631544 (9.2M)\n",
      "Saving to: ‘/g/scb/zaugg/deuner/SCENIC+/inputdata/bedToBigBed’\n",
      "\n",
      "/g/scb/zaugg/deuner 100%[===================>]   9.18M  5.09MB/s    in 1.8s    \n",
      "\n",
      "2023-05-28 10:51:03 (5.09 MB/s) - ‘/g/scb/zaugg/deuner/SCENIC+/inputdata/bedToBigBed’ saved [9631544/9631544]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Also download a the program bedToBigBed this will be used to generate files which can be uploaded to the UCSC genome browser\n",
    "!wget -O /g/scb/zaugg/deuner/SCENIC+/inputdata/bedToBigBed http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/bedToBigBed\n",
    "!chmod +x /g/scb/zaugg/deuner/SCENIC+/inputdata/bedToBigBed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7ce6e718-d5eb-4f73-83a9-3054bec13b23",
   "metadata": {},
   "outputs": [],
   "source": [
    "#only keep the first two columns of the PCA embedding in order to be able to visualize this in SCope\n",
    "#scplus_obj.dr_cell['GEX_X_pca'] = scplus_obj.dr_cell['GEX_X_pca'].iloc[:, 0:2]\n",
    "#scplus_obj.dr_cell['GEX_rep'] = scplus_obj.dr_cell['GEX_rep'].iloc[:, 0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5ff83897-8694-4f7b-bcdb-12d82752b238",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import ray\n",
    "# ray.shutdown()\n",
    "# ray.init()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "34f812d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-05-28 10:52:02,246 SCENIC+_wrapper INFO     /g/scb/zaugg/deuner/SCENIC+/tmp/combined/scenicplus folder already exists.\n",
      "2023-05-28 10:52:02,247 SCENIC+_wrapper INFO     Merging cistromes\n",
      "2023-05-28 11:02:06,754 SCENIC+_wrapper INFO     Getting search space\n",
      "2023-05-28 11:02:08,809 R2G          INFO     Downloading gene annotation from biomart dataset: hsapiens_gene_ensembl\n",
      "2023-05-28 11:02:21,881 R2G          INFO     Downloading chromosome sizes from: http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes\n",
      "2023-05-28 11:02:23,324 R2G          INFO     Extending promoter annotation to 10 bp upstream and 10 downstream\n",
      "2023-05-28 11:02:42,457 R2G          INFO     Extending search space to:\n",
      "            \t\t\t\t\t\t150000 bp downstream of the end of the gene.\n",
      "            \t\t\t\t\t\t150000 bp upstream of the start of the gene.\n",
      "2023-05-28 11:03:11,011 R2G          INFO     Intersecting with regions.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "join: Strand data from other will be added as strand data to self.\n",
      "If this is undesired use the flag apply_strand_suffix=False.\n",
      "To turn off the warning set apply_strand_suffix to True or False.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-05-28 11:03:12,322 R2G          INFO     Calculating distances from region to gene\n",
      "2023-05-28 11:08:23,454 R2G          INFO     Imploding multiple entries per region and gene\n",
      "2023-05-28 11:17:36,574 R2G          INFO     Done!\n",
      "2023-05-28 11:17:37,334 SCENIC+_wrapper INFO     Inferring region to gene relationships\n",
      "2023-05-28 11:17:37,976 R2G          INFO     Calculating region to gene importances, using GBM method\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-05-28 11:17:55,147\tINFO worker.py:1529 -- Started a local Ray instance. View the dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265 \u001b[39m\u001b[22m\n",
      "initializing: 100%|███████████████████████| 17522/17522 [38:30<00:00,  7.58it/s]\n",
      "Running using 24 cores: 100%|█████████████| 17522/17522 [10:29<00:00, 27.82it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-05-28 12:07:17,334 R2G          INFO     Took 2979.356848716736 seconds\n",
      "2023-05-28 12:07:17,338 R2G          INFO     Calculating region to gene correlation, using SR method\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-05-28 12:07:35,686\tINFO worker.py:1529 -- Started a local Ray instance. View the dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265 \u001b[39m\u001b[22m\n",
      "initializing:   2%|▍                        | 291/17522 [00:39<39:10,  7.33it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762371)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762371)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:   4%|▉                        | 680/17522 [01:27<35:08,  7.99it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:   4%|█                        | 724/17522 [01:33<34:01,  8.23it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:   8%|█▊                      | 1357/17522 [02:52<32:10,  8.38it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:   8%|█▉                      | 1411/17522 [02:58<32:52,  8.17it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:   8%|█▉                      | 1417/17522 [02:59<32:15,  8.32it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:   9%|██                      | 1496/17522 [03:09<32:15,  8.28it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:   9%|██▎                     | 1662/17522 [03:30<30:38,  8.62it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  11%|██▋                     | 1935/17522 [04:03<32:12,  8.07it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  13%|███                     | 2216/17522 [04:38<31:50,  8.01it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  13%|███                     | 2234/17522 [04:40<31:54,  7.98it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  13%|███▏                    | 2293/17522 [04:47<30:40,  8.27it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  13%|███▏                    | 2365/17522 [04:56<30:34,  8.26it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  16%|███▊                    | 2761/17522 [05:45<29:41,  8.28it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  17%|████                    | 2932/17522 [06:06<30:13,  8.05it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  25%|█████▉                  | 4311/17522 [08:56<26:05,  8.44it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  25%|█████▉                  | 4341/17522 [08:59<26:37,  8.25it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  25%|██████                  | 4437/17522 [09:11<25:51,  8.43it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  25%|██████                  | 4452/17522 [09:13<27:13,  8.00it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  26%|██████▏                 | 4547/17522 [09:25<27:17,  7.92it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  26%|██████▏                 | 4551/17522 [09:25<27:10,  7.96it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  26%|██████▎                 | 4606/17522 [09:32<26:48,  8.03it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  27%|██████▍                 | 4661/17522 [09:39<27:46,  7.72it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  27%|██████▍                 | 4735/17522 [10:11<26:55,  7.92it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  30%|███████                 | 5187/17522 [11:08<24:25,  8.41it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  30%|███████                 | 5198/17522 [11:09<25:57,  7.91it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  31%|███████▍                | 5386/17522 [11:32<25:23,  7.97it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  33%|███████▊                | 5695/17522 [12:11<24:35,  8.02it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  33%|███████▊                | 5739/17522 [12:16<25:19,  7.75it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  33%|███████▉                | 5820/17522 [12:26<24:16,  8.03it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  34%|████████▏               | 5981/17522 [12:46<23:20,  8.24it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  36%|████████▌               | 6244/17522 [13:19<23:27,  8.01it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  37%|████████▊               | 6435/17522 [13:42<22:04,  8.37it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  37%|████████▉               | 6546/17522 [13:56<20:46,  8.80it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  39%|█████████▎              | 6775/17522 [14:24<22:15,  8.05it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  43%|██████████▍             | 7605/17522 [16:06<19:55,  8.29it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  44%|██████████▌             | 7718/17522 [16:19<19:48,  8.25it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  46%|███████████             | 8033/17522 [16:58<19:03,  8.30it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  50%|███████████▉            | 8706/17522 [18:21<17:26,  8.43it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  50%|████████████            | 8764/17522 [18:28<18:15,  8.00it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  50%|████████████            | 8771/17522 [18:29<18:12,  8.01it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  50%|████████████            | 8827/17522 [18:36<18:02,  8.03it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  51%|████████████▏           | 8891/17522 [18:44<17:41,  8.13it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  52%|████████████▍           | 9112/17522 [19:11<17:18,  8.10it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  52%|████████████▌           | 9184/17522 [19:20<16:51,  8.24it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  54%|████████████▉           | 9434/17522 [20:21<17:41,  7.62it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  54%|████████████▉           | 9450/17522 [20:23<16:59,  7.91it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  55%|█████████████           | 9552/17522 [20:36<17:08,  7.75it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  55%|█████████████▎          | 9706/17522 [20:56<16:19,  7.98it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  58%|█████████████▎         | 10152/17522 [21:51<15:16,  8.04it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  59%|█████████████▌         | 10295/17522 [22:09<14:48,  8.13it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  60%|█████████████▊         | 10520/17522 [22:37<14:11,  8.22it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  60%|█████████████▉         | 10593/17522 [22:46<14:54,  7.75it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  62%|██████████████▎        | 10879/17522 [23:21<12:56,  8.56it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  63%|██████████████▍        | 11042/17522 [23:41<13:12,  8.18it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  63%|██████████████▌        | 11114/17522 [23:50<12:27,  8.57it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  65%|███████████████        | 11475/17522 [24:34<12:19,  8.18it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  70%|███████████████▉       | 12185/17522 [26:01<10:14,  8.69it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  71%|████████████████▎      | 12433/17522 [26:31<10:21,  8.19it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  72%|████████████████▌      | 12601/17522 [26:51<09:48,  8.36it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  72%|████████████████▌      | 12629/17522 [26:55<09:51,  8.27it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  73%|████████████████▋      | 12754/17522 [27:10<09:52,  8.05it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  73%|████████████████▊      | 12794/17522 [27:15<09:46,  8.06it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  75%|█████████████████▏     | 13124/17522 [27:55<08:50,  8.30it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  75%|█████████████████▎     | 13207/17522 [28:05<08:45,  8.21it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  75%|█████████████████▎     | 13223/17522 [28:07<08:56,  8.01it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  76%|█████████████████▌     | 13358/17522 [28:24<08:08,  8.53it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  78%|█████████████████▉     | 13629/17522 [28:57<07:58,  8.14it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  78%|█████████████████▉     | 13677/17522 [29:03<07:44,  8.27it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  79%|██████████████████     | 13766/17522 [29:14<07:31,  8.31it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  83%|███████████████████▏   | 14601/17522 [31:20<06:11,  7.86it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  84%|███████████████████▎   | 14743/17522 [31:37<05:54,  7.84it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  85%|███████████████████▌   | 14886/17522 [31:55<05:23,  8.14it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  85%|███████████████████▌   | 14920/17522 [31:59<05:18,  8.16it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762406)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  86%|███████████████████▋   | 15012/17522 [32:10<05:06,  8.19it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  89%|████████████████████▍  | 15612/17522 [33:24<03:42,  8.59it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762348)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  90%|████████████████████▋  | 15772/17522 [33:43<03:24,  8.54it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762379)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  91%|████████████████████▉  | 15975/17522 [34:08<03:09,  8.15it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762352)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  96%|██████████████████████ | 16822/17522 [35:52<01:23,  8.35it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762368)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing:  96%|██████████████████████▏| 16887/17522 [36:01<01:19,  8.02it/s]\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762362)\u001b[0m /g/scb/zaugg/deuner/miniconda3/envs/scenicplus/lib/python3.8/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "\u001b[2m\u001b[36m(_score_regions_to_single_gene_ray pid=1762362)\u001b[0m   warnings.warn(stats.ConstantInputWarning(warn_msg))\n",
      "initializing: 100%|███████████████████████| 17522/17522 [37:20<00:00,  7.82it/s]\n",
      "Running using 24 cores: 100%|████████████| 17522/17522 [01:08<00:00, 257.56it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-05-28 12:46:23,040 R2G          INFO     Took 2345.7002811431885 seconds\n",
      "2023-05-28 12:46:38,601 R2G          INFO     Done!\n",
      "2023-05-28 12:46:39,030 SCENIC+_wrapper INFO     Inferring TF to gene relationships\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-05-28 12:46:59,828\tINFO worker.py:1529 -- Started a local Ray instance. View the dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265 \u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-05-28 12:47:13,358 TF2G         INFO     Calculating TF to gene correlation, using GBM method\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "initializing:   1%|▏                      | 213/30768 [03:24<4:42:22,  1.80it/s]\u001b[2m\u001b[36m(raylet)\u001b[0m Spilled 4837 MiB, 2 objects, write throughput 474 MiB/s. Set RAY_verbose_spill_logs=0 to disable this message.\n",
      "\u001b[2m\u001b[36m(raylet)\u001b[0m Spilled 9675 MiB, 5 objects, write throughput 858 MiB/s.\n",
      "\u001b[2m\u001b[36m(raylet)\u001b[0m Spilled 19351 MiB, 11 objects, write throughput 1327 MiB/s.\n",
      "\u001b[2m\u001b[36m(raylet)\u001b[0m Spilled 24188 MiB, 14 objects, write throughput 1536 MiB/s.\n",
      "initializing:   1%|▏                      | 235/30768 [03:50<2:44:39,  3.09it/s]\u001b[2m\u001b[36m(raylet)\u001b[0m Spilled 53216 MiB, 33 objects, write throughput 1728 MiB/s.\n",
      "\u001b[2m\u001b[33m(raylet)\u001b[0m [2023-05-28 12:51:06,022 E 1973851 1973874] (raylet) file_system_monitor.cc:105: /tmp/ray/session_2023-05-28_12-46-41_694128_1511726 is over 95% full, available space: 5491822592; capacity: 129866661888. Object creation will fail if spilling is required.\n",
      "initializing:   1%|▏                      | 243/30768 [03:55<3:10:28,  2.67it/s]\u001b[2m\u001b[33m(raylet)\u001b[0m [2023-05-28 12:51:16,042 E 1973851 1973874] (raylet) file_system_monitor.cc:105: /tmp/ray/session_2023-05-28_12-46-41_694128_1511726 is over 95% full, available space: 5073223680; capacity: 129866661888. Object creation will fail if spilling is required.\n",
      "\u001b[2m\u001b[33m(raylet)\u001b[0m [2023-05-28 12:51:26,057 E 1973851 1973874] (raylet) file_system_monitor.cc:105: /tmp/ray/session_2023-05-28_12-46-41_694128_1511726 is over 95% full, available space: 5073223680; capacity: 129866661888. Object creation will fail if spilling is required.\n",
      "initializing:   1%|▏                      | 258/30768 [04:21<3:34:45,  2.37it/s]\u001b[2m\u001b[33m(raylet)\u001b[0m [2023-05-28 12:51:36,072 E 1973851 1973874] (raylet) file_system_monitor.cc:105: /tmp/ray/session_2023-05-28_12-46-41_694128_1511726 is over 95% full, available space: 5073223680; capacity: 129866661888. Object creation will fail if spilling is required.\n",
      "initializing:   1%|▏                      | 260/30768 [04:23<6:01:19,  1.41it/s]\u001b[2m\u001b[33m(raylet)\u001b[0m [2023-05-28 12:51:46,090 E 1973851 1973874] (raylet) file_system_monitor.cc:105: /tmp/ray/session_2023-05-28_12-46-41_694128_1511726 is over 95% full, available space: 5073199104; capacity: 129866661888. Object creation will fail if spilling is required.\n",
      "initializing:   1%|▏                      | 261/30768 [04:41<9:08:43,  1.08s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Local disk is full\n",
      "The object cannot be created because the local object store is full and the local disk's utilization is over capacity (95% by default).Tip: Use `df` on this node to check disk usage and `ray memory` to check object store memory usage.\n",
      "2023-05-28 12:52:47,860 TF2G         INFO     Took 334.49985218048096 seconds\n",
      "2023-05-28 12:52:47,861 TF2G         INFO     Adding correlation coefficients to adjacencies.\n"
     ]
    },
    {
     "ename": "UnboundLocalError",
     "evalue": "local variable 'tfs_to_genes' referenced before assignment",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[17], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mscenicplus\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mwrappers\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mrun_scenicplus\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m run_scenicplus\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m# try:\u001b[39;00m\n\u001b[0;32m----> 4\u001b[0m \u001b[43mrun_scenicplus\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      5\u001b[0m \u001b[43m        \u001b[49m\u001b[43mscplus_obj\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mscplus_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      6\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvariable\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mGEX_celltype\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[43m        \u001b[49m\u001b[43mspecies\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mhsapiens\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[43m        \u001b[49m\u001b[43massembly\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mhg38\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      9\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtf_file\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m/g/scb/zaugg/deuner/SCENIC+/inputdata/utoronto_human_tfs_v_1.01.txt\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     10\u001b[0m \u001b[43m        \u001b[49m\u001b[43msave_path\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtmp_dir\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mscenicplus\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     11\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbiomart_host\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mbiomart_host\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     12\u001b[0m \u001b[43m        \u001b[49m\u001b[43mupstream\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m1000\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m150000\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     13\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdownstream\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m1000\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m150000\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     14\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcalculate_TF_eGRN_correlation\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     15\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcalculate_DEGs_DARs\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     16\u001b[0m \u001b[43m        \u001b[49m\u001b[43mexport_to_loom_file\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     17\u001b[0m \u001b[43m        \u001b[49m\u001b[43mexport_to_UCSC_file\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     18\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpath_bedToBigBed\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m/g/scb/zaugg/deuner/SCENIC+/inputdata\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     19\u001b[0m \u001b[43m        \u001b[49m\u001b[43mn_cpu\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m24\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     20\u001b[0m \u001b[43m        \u001b[49m\u001b[43m_temp_dir\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/g/scb/zaugg/deuner/SCENIC+/scenicplus/src/scenicplus/wrappers/run_scenicplus.py:151\u001b[0m, in \u001b[0;36mrun_scenicplus\u001b[0;34m(scplus_obj, variable, species, assembly, tf_file, save_path, biomart_host, upstream, downstream, region_ranking, gene_ranking, simplified_eGRN, calculate_TF_eGRN_correlation, calculate_DEGs_DARs, export_to_loom_file, export_to_UCSC_file, tree_structure, path_bedToBigBed, n_cpu, _temp_dir, **kwargs)\u001b[0m\n\u001b[1;32m    149\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTF2G_adj\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m scplus_obj\u001b[38;5;241m.\u001b[39muns\u001b[38;5;241m.\u001b[39mkeys():\n\u001b[1;32m    150\u001b[0m     log\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mInferring TF to gene relationships\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m--> 151\u001b[0m     \u001b[43mcalculate_TFs_to_genes_relationships\u001b[49m\u001b[43m(\u001b[49m\u001b[43mscplus_obj\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m    152\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mtf_file\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mtf_file\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    153\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mray_n_cpu\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mn_cpu\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m    154\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mGBM\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    155\u001b[0m \u001b[43m                    \u001b[49m\u001b[43m_temp_dir\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m_temp_dir\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    156\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mkey\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mTF2G_adj\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    157\u001b[0m \u001b[43m                    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    159\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124meRegulons\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m scplus_obj\u001b[38;5;241m.\u001b[39muns\u001b[38;5;241m.\u001b[39mkeys():\n\u001b[1;32m    160\u001b[0m     log\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mBuild eGRN\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m/g/scb/zaugg/deuner/SCENIC+/scenicplus/src/scenicplus/TF_to_gene.py:332\u001b[0m, in \u001b[0;36mcalculate_TFs_to_genes_relationships\u001b[0;34m(scplus_obj, tf_file, method, ray_n_cpu, key, **kwargs)\u001b[0m\n\u001b[1;32m    330\u001b[0m start_time \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime()\n\u001b[1;32m    331\u001b[0m log\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mAdding correlation coefficients to adjacencies.\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m--> 332\u001b[0m adj \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat(\u001b[43mtfs_to_genes\u001b[49m)\u001b[38;5;241m.\u001b[39msort_values(by\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mimportance\u001b[39m\u001b[38;5;124m'\u001b[39m, ascending\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m    333\u001b[0m ex_matrix \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(\n\u001b[1;32m    334\u001b[0m     scplus_obj\u001b[38;5;241m.\u001b[39mX_EXP, index\u001b[38;5;241m=\u001b[39mscplus_obj\u001b[38;5;241m.\u001b[39mcell_names, columns\u001b[38;5;241m=\u001b[39mscplus_obj\u001b[38;5;241m.\u001b[39mgene_names)\n\u001b[1;32m    335\u001b[0m adj \u001b[38;5;241m=\u001b[39m _add_correlation(adj, ex_matrix)\n",
      "\u001b[0;31mUnboundLocalError\u001b[0m: local variable 'tfs_to_genes' referenced before assignment"
     ]
    }
   ],
   "source": [
    "# Run the analysis\n",
    "from scenicplus.wrappers.run_scenicplus import run_scenicplus\n",
    "# try:\n",
    "run_scenicplus(\n",
    "        scplus_obj = scplus_obj,\n",
    "        variable = ['GEX_celltype'],\n",
    "        species = 'hsapiens',\n",
    "        assembly = 'hg38',\n",
    "        tf_file = '/g/scb/zaugg/deuner/SCENIC+/inputdata/utoronto_human_tfs_v_1.01.txt',\n",
    "        save_path = os.path.join(tmp_dir, 'scenicplus'),\n",
    "        biomart_host = biomart_host,\n",
    "        upstream = [1000, 150000],\n",
    "        downstream = [1000, 150000],\n",
    "        calculate_TF_eGRN_correlation = True,\n",
    "        calculate_DEGs_DARs = True,\n",
    "        export_to_loom_file = True,\n",
    "        export_to_UCSC_file = True,\n",
    "        path_bedToBigBed = '/g/scb/zaugg/deuner/SCENIC+/inputdata',\n",
    "        n_cpu = 24,\n",
    "        _temp_dir = None)#'/g/scb/zaugg/deuner/ray_spill')\n",
    "# except Exception as e:\n",
    "#     #in case of failure, still save the object\n",
    "#     dill.dump(scplus_obj, open(os.path.join(out_dir, '/scplus_obj.pkl'), 'wb'), protocol=-1)\n",
    "#     raise(e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8be6ee7",
   "metadata": {},
   "source": [
    "## Downstream Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d26bbf06",
   "metadata": {},
   "source": [
    "### Simplifying and filtering SCENIC+ output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dccac685",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scenicplus.preprocessing.filtering import apply_std_filtering_to_eRegulons\n",
    "apply_std_filtering_to_eRegulons(scplus_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1af4f617",
   "metadata": {},
   "outputs": [],
   "source": [
    "scplus_obj.uns['eRegulon_metadata_filtered'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c30818eb",
   "metadata": {},
   "source": [
    "### eRegulon enrichment scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91054b11",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scenicplus.eregulon_enrichment import score_eRegulons\n",
    "region_ranking = dill.load(open(os.path.join(out_dir, 'scenicplus/region_ranking.pkl'), 'rb')) #load ranking calculated using the wrapper function\n",
    "gene_ranking = dill.load(open(os.path.join(out_dir, 'scenicplus/gene_ranking.pkl'), 'rb')) #load ranking calculated using the wrapper function\n",
    "score_eRegulons(scplus_obj,\n",
    "                ranking = region_ranking,\n",
    "                eRegulon_signatures_key = 'eRegulon_signatures_filtered',\n",
    "                key_added = 'eRegulon_AUC_filtered',\n",
    "                enrichment_type= 'region',\n",
    "                auc_threshold = 0.05,\n",
    "                normalize = False,\n",
    "                n_cpu = 5)\n",
    "score_eRegulons(scplus_obj,\n",
    "                gene_ranking,\n",
    "                eRegulon_signatures_key = 'eRegulon_signatures_filtered',\n",
    "                key_added = 'eRegulon_AUC_filtered',\n",
    "                enrichment_type = 'gene',\n",
    "                auc_threshold = 0.05,\n",
    "                normalize= False,\n",
    "                n_cpu = 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f0cc7a8",
   "metadata": {},
   "source": [
    "### eRegulon dimensionality reduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f1ef360",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scenicplus.dimensionality_reduction import run_eRegulons_tsne, run_eRegulons_umap\n",
    "run_eRegulons_umap(\n",
    "    scplus_obj = scplus_obj,\n",
    "    auc_key = 'eRegulon_AUC_filtered',\n",
    "    reduction_name = 'eRegulons_UMAP', #overwrite previously calculated UMAP\n",
    ")\n",
    "run_eRegulons_tsne(\n",
    "    scplus_obj = scplus_obj,\n",
    "    auc_key = 'eRegulon_AUC_filtered',\n",
    "    reduction_name = 'eRegulons_tSNE', #overwrite previously calculated tSNE\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f208ea75",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize it\n",
    "from scenicplus.dimensionality_reduction import plot_metadata_given_ax\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "%matplotlib inline\n",
    "\n",
    "#specify color_dictionary\n",
    "\n",
    "color_dict = {\n",
    "    'neuron': \"#065143\",\n",
    "    'hiPSC': \"#70B77E\",\n",
    "    'microglia': \"#E0A890\",\n",
    "    'diff.state': \"#053C5E\" \n",
    "}\n",
    "\n",
    "fig, axs = plt.subplots(ncols=2, figsize = (16, 8))\n",
    "plot_metadata_given_ax(\n",
    "    scplus_obj=scplus_obj,\n",
    "    ax = axs[0],\n",
    "    reduction_name = 'eRegulons_UMAP',\n",
    "    variable = 'GEX_celltype', #note the GEX_ prefix, this metadata originated from the gene expression metadata (on which we did the cell type annotation before)\n",
    "    color_dictionary={'GEX_celltype': color_dict}\n",
    ")\n",
    "plot_metadata_given_ax(\n",
    "    scplus_obj=scplus_obj,\n",
    "    ax = axs[1],\n",
    "    reduction_name = 'eRegulons_tSNE',\n",
    "    variable = 'GEX_celltype', #note the GEX_ prefix, this metadata originated from the gene expression metadata (on which we did the cell type annotation before)\n",
    "    color_dictionary={'GEX_celltype': color_dict}\n",
    ")\n",
    "fig.tight_layout()\n",
    "sns.despine(ax = axs[0]) #remove top and right edge of axis border\n",
    "sns.despine(ax = axs[1]) #remove top and right edge of axis border\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71ffb3b4",
   "metadata": {},
   "source": [
    "### plot the activity / expression of an eRegulon on the dimensionality reduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f8e8bea",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scenicplus.dimensionality_reduction import plot_eRegulon\n",
    "plot_eRegulon(\n",
    "    scplus_obj = scplus_obj,\n",
    "    reduction_name = 'eRegulons_tSNE',\n",
    "    selected_regulons = ['POU4F3', 'KLF12', 'POU4F1', 'CUX2', 'ONECUT3'],\n",
    "    scale = True,\n",
    "    auc_key = 'eRegulon_AUC_filtered')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67277c0f",
   "metadata": {},
   "source": [
    "### dotplot-heatmap\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68bdd74f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We first generate pseudobulk gene expression and region accessibility data, per celltype, to limit the amount of noise for the correlation calculation.\n",
    "from scenicplus.cistromes import TF_cistrome_correlation, generate_pseudobulks\n",
    "\n",
    "generate_pseudobulks(\n",
    "        scplus_obj = scplus_obj,\n",
    "        variable = 'GEX_celltype',\n",
    "        auc_key = 'eRegulon_AUC_filtered',\n",
    "        signature_key = 'Gene_based')\n",
    "generate_pseudobulks(\n",
    "        scplus_obj = scplus_obj,\n",
    "        variable = 'GEX_celltype',\n",
    "        auc_key = 'eRegulon_AUC_filtered',\n",
    "        signature_key = 'Region_based')\n",
    "\n",
    "TF_cistrome_correlation(\n",
    "            scplus_obj,\n",
    "            use_pseudobulk = True,\n",
    "            variable = 'GEX_celltype',\n",
    "            auc_key = 'eRegulon_AUC_filtered',\n",
    "            signature_key = 'Gene_based',\n",
    "            out_key = 'filtered_gene_based')\n",
    "TF_cistrome_correlation(\n",
    "            scplus_obj,\n",
    "            use_pseudobulk = True,\n",
    "            variable = 'GEX_celltype',\n",
    "            auc_key = 'eRegulon_AUC_filtered',\n",
    "            signature_key = 'Region_based',\n",
    "            out_key = 'filtered_region_based')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b4a1ca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "scplus_obj.uns['TF_cistrome_correlation']['filtered_region_based'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c722cb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's visualize these correlations in a scatter plot and select eRegulons for which the correlaiton coefficient is above 0.70 or below -0.75\n",
    "import numpy as np\n",
    "n_targets = [int(x.split('(')[1].replace('r)', '')) for x in scplus_obj.uns['TF_cistrome_correlation']['filtered_region_based']['Cistrome']]\n",
    "rho = scplus_obj.uns['TF_cistrome_correlation']['filtered_region_based']['Rho'].to_list()\n",
    "adj_pval = scplus_obj.uns['TF_cistrome_correlation']['filtered_region_based']['Adjusted_p-value'].to_list()\n",
    "\n",
    "thresholds = {\n",
    "        'rho': [-0.75, 0.70],\n",
    "        'n_targets': 0\n",
    "}\n",
    "import seaborn as sns\n",
    "fig, ax = plt.subplots(figsize = (10, 5))\n",
    "sc = ax.scatter(rho, n_targets, c = -np.log10(adj_pval), s = 5)\n",
    "ax.set_xlabel('Correlation coefficient')\n",
    "ax.set_ylabel('nr. target regions')\n",
    "#ax.hlines(y = thresholds['n_targets'], xmin = min(rho), xmax = max(rho), color = 'black', ls = 'dashed', lw = 1)\n",
    "ax.vlines(x = thresholds['rho'], ymin = 0, ymax = max(n_targets), color = 'black', ls = 'dashed', lw = 1)\n",
    "ax.text(x = thresholds['rho'][0], y = max(n_targets), s = str(thresholds['rho'][0]))\n",
    "ax.text(x = thresholds['rho'][1], y = max(n_targets), s = str(thresholds['rho'][1]))\n",
    "sns.despine(ax = ax)\n",
    "fig.colorbar(sc, label = '-log10(adjusted_pvalue)', ax = ax)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3fc08d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_cistromes = scplus_obj.uns['TF_cistrome_correlation']['filtered_region_based'].loc[\n",
    "        np.logical_or(\n",
    "                scplus_obj.uns['TF_cistrome_correlation']['filtered_region_based']['Rho'] > thresholds['rho'][1],\n",
    "                scplus_obj.uns['TF_cistrome_correlation']['filtered_region_based']['Rho'] < thresholds['rho'][0]\n",
    "        )]['Cistrome'].to_list()\n",
    "selected_eRegulons = [x.split('_(')[0] for x in selected_cistromes]\n",
    "selected_eRegulons_gene_sig = [\n",
    "        x for x in scplus_obj.uns['eRegulon_signatures_filtered']['Gene_based'].keys()\n",
    "        if x.split('_(')[0] in selected_eRegulons]\n",
    "selected_eRegulons_region_sig = [\n",
    "        x for x in scplus_obj.uns['eRegulon_signatures_filtered']['Region_based'].keys()\n",
    "        if x.split('_(')[0] in selected_eRegulons]\n",
    "#save the results in the scenicplus object\n",
    "scplus_obj.uns['selected_eRegulon'] = {'Gene_based': selected_eRegulons_gene_sig, 'Region_based': selected_eRegulons_region_sig}\n",
    "print(f'selected: {len(selected_eRegulons_gene_sig)} eRegulons')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "383e5a0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save these changes we have made to the scenicplus_obj\n",
    "dill.dump(scplus_obj, open(os.path.join(out_dir, 'scenicplus/scplus_obj.pkl'), 'wb'), protocol=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c07ea6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the heatmap-dotplot\n",
    "from scenicplus.plotting.dotplot import heatmap_dotplot\n",
    "heatmap_dotplot(\n",
    "        scplus_obj = scplus_obj,\n",
    "        size_matrix = scplus_obj.uns['eRegulon_AUC_filtered']['Region_based'], #specify what to plot as dot sizes, target region enrichment in this case\n",
    "        color_matrix = scplus_obj.to_df('EXP'), #specify  what to plot as colors, TF expression in this case\n",
    "        scale_size_matrix = True,\n",
    "        scale_color_matrix = True,\n",
    "        group_variable = 'GEX_celltype',\n",
    "        subset_eRegulons = scplus_obj.uns['selected_eRegulon']['Gene_based'],\n",
    "        figsize = (5, 20),\n",
    "        orientation = 'vertical')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61f706d5",
   "metadata": {},
   "source": [
    "### Overlap of predicted target regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a888121",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate the RSS for the target regions of the selected eRegulons.\n",
    "from scenicplus.RSS import *\n",
    "regulon_specificity_scores(\n",
    "        scplus_obj,\n",
    "        variable = 'GEX_celltype',\n",
    "        auc_key = 'eRegulon_AUC_filtered',\n",
    "        signature_keys = ['Region_based'],\n",
    "        selected_regulons = [x for x in scplus_obj.uns['selected_eRegulon']['Region_based'] if '-' not in x],\n",
    "        out_key_suffix = '_filtered')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7a4d3bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# visualize the RSS values using a scatter plot\n",
    "plot_rss(scplus_obj, 'GEX_celltype_filtered', num_columns=2, top_n=10, figsize = (5, 10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72d13017",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select the top 10 eRegulons per cell type\n",
    "flat_list = lambda t: [item for sublist in t for item in sublist]\n",
    "selected_markers = list(set(flat_list(\n",
    "    [scplus_obj.uns['RSS']['GEX_celltype_filtered'].loc[celltype].sort_values(ascending = False).head(10).index.to_list()\n",
    "    for celltype in scplus_obj.uns['RSS']['GEX_celltype_filtered'].index])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf1cccd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scenicplus.plotting.correlation_plot import *\n",
    "\n",
    "region_intersetc_data, Z = jaccard_heatmap(\n",
    "        scplus_obj,\n",
    "        method = 'intersect',\n",
    "        gene_or_region_based = 'Region_based',\n",
    "        use_plotly = False,\n",
    "        selected_regulons = selected_markers,\n",
    "        signature_key = 'eRegulon_signatures_filtered',\n",
    "        figsize = (10, 10), return_data = True, vmax = 0.5, cmap = 'plasma')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8430f5f3",
   "metadata": {},
   "source": [
    "### Plotting a Network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a923811",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycisTopic.diff_features import find_highly_variable_features\n",
    "hvr = find_highly_variable_features(scplus_obj.to_df('ACC').loc[list(set(scplus_obj.uns['eRegulon_metadata_filtered']['Region']))], n_top_features=1000, plot = False)\n",
    "hvg = find_highly_variable_features(scplus_obj.to_df('EXP')[list(set(scplus_obj.uns['eRegulon_metadata_filtered']['Gene']))].T, n_top_features=1000, plot = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "781e5000",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scenicplus.networks import create_nx_tables, create_nx_graph, plot_networkx, export_to_cytoscape\n",
    "nx_tables = create_nx_tables(\n",
    "    scplus_obj = scplus_obj,\n",
    "    eRegulon_metadata_key ='eRegulon_metadata_filtered',\n",
    "    subset_eRegulons = ['PAX5', 'EBF1', 'POU2AF1'],\n",
    "    subset_regions = hvr,\n",
    "    subset_genes = hvg,\n",
    "    add_differential_gene_expression = True,\n",
    "    add_differential_region_accessibility = True,\n",
    "    differential_variable = ['GEX_celltype'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78030306",
   "metadata": {},
   "outputs": [],
   "source": [
    "G, pos, edge_tables, node_tables = create_nx_graph(nx_tables,\n",
    "                   use_edge_tables = ['TF2R','R2G'],\n",
    "                   color_edge_by = {'TF2R': {'variable' : 'TF', 'category_color' : {'PAX5': 'Orange', 'EBF1': 'Purple', 'POU2AF1': 'Red'}},\n",
    "                                    'R2G': {'variable' : 'R2G_rho', 'continuous_color' : 'viridis', 'v_min': -1, 'v_max': 1}},\n",
    "                   transparency_edge_by =  {'R2G': {'variable' : 'R2G_importance', 'min_alpha': 0.1, 'v_min': 0}},\n",
    "                   width_edge_by = {'R2G': {'variable' : 'R2G_importance', 'max_size' :  1.5, 'min_size' : 1}},\n",
    "                   color_node_by = {'TF': {'variable': 'TF', 'category_color' : {'PAX5': 'Orange', 'EBF1': 'Purple', 'POU2AF1': 'Red'}},\n",
    "                                    'Gene': {'variable': 'GEX_celltype_Log2FC_B_cells_1', 'continuous_color' : 'bwr'},\n",
    "                                    'Region': {'variable': 'GEX_celltype_Log2FC_B_cells_1', 'continuous_color' : 'viridis'}},\n",
    "                   transparency_node_by =  {'Region': {'variable' : 'GEX_celltype_Log2FC_B_cells_1', 'min_alpha': 0.1},\n",
    "                                    'Gene': {'variable' : 'GEX_celltype_Log2FC_B_cells_1', 'min_alpha': 0.1}},\n",
    "                   size_node_by = {'TF': {'variable': 'fixed_size', 'fixed_size': 30},\n",
    "                                    'Gene': {'variable': 'fixed_size', 'fixed_size': 15},\n",
    "                                    'Region': {'variable': 'fixed_size', 'fixed_size': 10}},\n",
    "                   shape_node_by = {'TF': {'variable': 'fixed_shape', 'fixed_shape': 'ellipse'},\n",
    "                                    'Gene': {'variable': 'fixed_shape', 'fixed_shape': 'ellipse'},\n",
    "                                    'Region': {'variable': 'fixed_shape', 'fixed_shape': 'diamond'}},\n",
    "                   label_size_by = {'TF': {'variable': 'fixed_label_size', 'fixed_label_size': 20.0},\n",
    "                                    'Gene': {'variable': 'fixed_label_size', 'fixed_label_size': 10.0},\n",
    "                                    'Region': {'variable': 'fixed_label_size', 'fixed_label_size': 0.0}},\n",
    "                   layout='kamada_kawai_layout',\n",
    "                   scale_position_by=250)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6d02bfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(10,10))\n",
    "plot_networkx(G, pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b2d1c1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "export_to_cytoscape(G, pos, out_file = os.path.join(out_dir, 'scenicplus/network_combined.cys'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sp",
   "language": "python",
   "name": "sp"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
